<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>6&nbsp; Scanning the genome for regions of interest – **[fmicompbio/footprintR]{.githubpkg}** book</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./grouping-reads.html" rel="next">
<link href="./nucleosomes.html" rel="prev">
<link href="./images/footprintR_logo.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-6082b3b653f69b4bc1378a5897c1dc6c.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./scanning.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Scanning the genome for regions of interest</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./"><strong></strong></a><strong><a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a></strong> book 
        <div class="sidebar-tools-main">
    <a href="https://github.com/fmicompbio/footprintR_book/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./reading-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Reading data from files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qc-filtering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality control and filtering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visualize-regions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualizing regions with <code>plotRegion</code></span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./nucleosomes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Nucleosome analyses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scanning.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Scanning the genome for regions of interest</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./grouping-reads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Grouping reads</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./multiple-modalities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Combining and comparing multiple data modalities</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#genome-wide-dmr-scanning-between-samples" id="toc-genome-wide-dmr-scanning-between-samples" class="nav-link active" data-scroll-target="#genome-wide-dmr-scanning-between-samples"><span class="header-section-number">6.1</span> Genome-wide DMR scanning (between samples)</a>
  <ul class="collapse">
  <li><a href="#choosing-suitable-values-for-thresholds-and-smoothing-parameters" id="toc-choosing-suitable-values-for-thresholds-and-smoothing-parameters" class="nav-link" data-scroll-target="#choosing-suitable-values-for-thresholds-and-smoothing-parameters"><span class="header-section-number">6.1.1</span> Choosing suitable values for thresholds and smoothing parameters</a></li>
  </ul></li>
  <li><a href="#genome-wide-estimation-of-cpg-methylation-levels-within-samples" id="toc-genome-wide-estimation-of-cpg-methylation-levels-within-samples" class="nav-link" data-scroll-target="#genome-wide-estimation-of-cpg-methylation-levels-within-samples"><span class="header-section-number">6.2</span> Genome-wide estimation of CpG methylation levels (within samples)</a></li>
  <li><a href="#annotation-of-predefined-windows" id="toc-annotation-of-predefined-windows" class="nav-link" data-scroll-target="#annotation-of-predefined-windows"><span class="header-section-number">6.3</span> Annotation of predefined windows</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="header-section-number">6.4</span> Session info</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/fmicompbio/footprintR_book/edit/main/scanning.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/fmicompbio/footprintR_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-scanning" class="quarto-section-identifier"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Scanning the genome for regions of interest</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a> provides a flexible scanning framework, which can be used to partition the genome (or a subset thereof) into windows and calculate a custom score for each window. Optionally, neighboring windows with similar scores can then be merged into larger signal regions. In this chapter, we show some examples of how this framework can be used to scan the genome for various types of signals of interest.</p>
<p>The figure below illustrates the general idea of the scanning framework.</p>
<div id="fig-scan-framework" class="quarto-float quarto-figure quarto-figure-center anchored" alt="Schematic of the footprintR genome scanning framework." height="600">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-scan-framework-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="images/scan-framework-v2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;6.1: "><img src="images/scan-framework-v2.png" id="fig-scan-framework" alt="Schematic of the footprintR genome scanning framework." height="600" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-scan-framework-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.1
</figcaption>
</figure>
</div>
<p>The first step is to define the windows and calculate some quantity of interest for each of them, using the <a href="https://fmicompbio.github.io/footprintR/reference/quantifyWindowsInRegion.html"><code>quantifyWindowsInRegion</code></a> function. Next, these quantities are passed to a <em>score function</em>, which summarizes the input data into a score for each window - this can, for example, perform a differential methylation analysis based on the counts of methylated and unmethylated nucleotides obtained from <a href="https://fmicompbio.github.io/footprintR/reference/quantifyWindowsInRegion.html"><code>quantifyWindowsInRegion</code></a>. Finally, the window scores are processed using <a href="https://fmicompbio.github.io/footprintR/reference/processWindowScores.html"><code>processWindowScores</code></a>. This function allows neighboring windows with similar scores to be merged into larger signal regions. The user can choose to either call the functions directly, or use the wrapper function <a href="https://fmicompbio.github.io/footprintR/reference/scanForHighScoringRegions.html"><code>scanForHighScoringRegions</code></a>.</p>
<p>Thanks to the modularity, the scanning framework is flexible and can accommodate many different use cases by selecting a suitable quantification function, scoring function and aggregation of the scores. The table below lists some of the applications that can be accommodated with the functions that are provided within <a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a>. However, the user can also define their own functions for specific use cases.</p>
<div id="fig-scan-usecases" class="quarto-float quarto-figure quarto-figure-center anchored" alt="Table of use cases that can be addressed using the genome scanning framework in footprintR." height="300">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-scan-usecases-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="images/scan-useCases.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;6.2: "><img src="images/scan-useCases.png" id="fig-scan-usecases" alt="Table of use cases that can be addressed using the genome scanning framework in footprintR." height="300" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-scan-usecases-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.2
</figcaption>
</figure>
</div>
<p>Below we will show several examples of using the scanning framework for different purposes.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>BSgenomeName <span class="ot">&lt;-</span> <span class="st">"BSgenome.Mmusculus.GENCODE.GRCm39.gencodeM34"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(footprintR)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenomeName, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Hmisc)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Load genome</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>gnm <span class="ot">&lt;-</span> <span class="fu">get</span>(BSgenomeName)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(gnm) <span class="ot">&lt;-</span> <span class="st">"mm39"</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Define parallelization backend</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>bpParam <span class="ot">&lt;-</span> <span class="fu">MulticoreParam</span>(<span class="at">workers =</span> <span class="dv">24</span>L, <span class="at">RNGseed =</span> <span class="dv">42</span>L)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="genome-wide-dmr-scanning-between-samples" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="genome-wide-dmr-scanning-between-samples"><span class="header-section-number">6.1</span> Genome-wide DMR scanning (between samples)</h2>
<p>The first use case involves scanning for differentially methylated regions (DMRs) between two conditions, based on 6mA data. In our case, we will compare wild-type mouse ES cells to ‘TKO’ mouse ES cells where the three DNA methyltransferases (Dnmts) have been knocked out. For time reasons, we limit the scanning to chromosome 19 - the most efficient way of performing a genome-wide analysis is typically to parallelize over chromosomes (e.g.&nbsp;by replacing the <code>lapply</code> call in the code below by a parallelized version).</p>
<p>We tile the genome into 30bp windows (<code>windowSize</code>), shifted successively by 15bp (<code>windowStep</code>) to achieve a high resolution and minimize the risk of splitting a region of interest in the middle (thereby potentially losing the signal). We then use the <a href="https://fmicompbio.github.io/footprintR/reference/sumNmodNvalid.html"><code>sumNmodNvalid</code></a> function to count the number of modified As for each window, as well as the total number of As. The classification of nucleotides is obtained by thresholding the modification probability using the <code>modProbThreshold</code> value (0.5). Given the modified and total count for each sample, we next use the <a href="https://fmicompbio.github.io/footprintR/reference/getDifferentiallyModifiedWindows.html"><code>getDifferentiallyModifiedWindows</code></a> score function to perform the differential methylation analysis for each window. This will return several statistics, including the <code>dirNegLog10PValue</code>, which is obtained by calculating <code>-log10(p-value)</code> and multiplying with the sign of the log-fold change. We will use this statistic (<code>scoreCol</code>) as the input for the final processing of the windows. Briefly, the statistics will be smoothed over neighboring windows (the degree of smoothing is controlled by the <code>minperiod</code> argument) and subsequently thresholded (<code>thresh</code> argument). Neighboring windows where the statistic exceeds this threshold will be fused into a larger region of interest, unless they are separated by more than 50 nucleotides (<code>maxGap</code>).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define bam files to include</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>bamfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"data/mESC_wt_6mA_rep1.bam"</span>, <span class="st">"data/mESC_wt_6mA_rep2.bam"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>              <span class="st">"data/mESC_TKO_6mA_rep1.bam"</span>, <span class="st">"data/mESC_TKO_6mA_rep2.bam"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bamfiles) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.bam"</span>, <span class="st">""</span>, <span class="fu">basename</span>(bamfiles))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># define sample annotation table</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>sampleAnnot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(bamfiles),</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">group =</span> <span class="fu">factor</span>(<span class="fu">str_extract</span>(<span class="fu">names</span>(bamfiles), <span class="st">"wt|TKO"</span>), </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"wt"</span>, <span class="st">"TKO"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run scanning of chromosome 19</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>chrlens <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(gnm)[<span class="st">"chr19"</span>]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>dmrL <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(chrlens), <span class="cf">function</span>(i) {</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scanForHighScoringRegions</span>(</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">bamfiles =</span> bamfiles,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">sampleAnnot =</span> sampleAnnot,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">chromosomeLengths =</span> chrlens[i],</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">quantFunction =</span> <span class="st">"sumNmodNvalid"</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreFunction =</span> <span class="st">"getDifferentiallyModifiedWindows"</span>,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">modbase =</span> <span class="st">"a"</span>,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">modProbThreshold =</span> <span class="fl">0.5</span>,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">tileSize =</span> <span class="fl">2e+06</span>,</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm),</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceReference =</span> gnm,</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowMode =</span> <span class="st">"fixed"</span>,</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowSize =</span> <span class="dv">30</span>,</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowStep =</span> <span class="dv">15</span>,</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreCol =</span> <span class="st">"dirNegLog10PValue"</span>,</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreAction =</span> <span class="st">"smoothFuse"</span>,</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">thresh =</span> <span class="dv">3</span>,</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">minperiod =</span> <span class="dv">3</span>,</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">maxGap =</span> <span class="dv">50</span>,</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">BPPARAM =</span> bpParam,</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Each element of the resulting list (in this case, there is only one) contains the inferred DMRs for one chromosome.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(dmrL[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32498</code></pre>
</div>
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dmrL[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 5 metadata columns:
      seqnames          ranges strand | dirNegLog10PValueThresh
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |               &lt;numeric&gt;
  [1]    chr19 3051451-3051480      * |                 3.02470
  [2]    chr19 3051841-3051915      * |                 3.49924
  [3]    chr19 3052036-3052110      * |                 5.05907
  [4]    chr19 3052591-3052650      * |                 3.72814
  [5]    chr19 3053191-3053265      * |                 3.83881
  [6]    chr19 3053386-3053445      * |                 3.86745
      numWindowsThresh direction dirNegLog10PValue numWindows
             &lt;integer&gt;  &lt;factor&gt;         &lt;numeric&gt;  &lt;integer&gt;
  [1]                1  positive           3.02470          1
  [2]                4  positive           3.49924          4
  [3]                4  positive           5.05907          4
  [4]                3  positive           3.72814          3
  [5]                4  positive           3.83881          4
  [6]                3  positive           3.86745          3
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dmrL[[<span class="dv">1</span>]]<span class="sc">$</span>direction)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
positive negative 
   32490        8 </code></pre>
</div>
</div>
<p>We can visualize one of the top hits using <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>reg <span class="ot">&lt;-</span> dmrL[[<span class="dv">1</span>]][<span class="fu">which.max</span>(<span class="fu">abs</span>(dmrL[[<span class="dv">1</span>]]<span class="sc">$</span>dirNegLog10PValue))]</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(<span class="at">bamfiles =</span> bamfiles, <span class="at">sampleAnnot =</span> sampleAnnot, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">regions =</span> reg <span class="sc">+</span> <span class="dv">500</span>, <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">sequenceReference =</span> gnm, <span class="at">trim =</span> <span class="cn">TRUE</span>, </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(se)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(se, <span class="at">region =</span> reg <span class="sc">+</span> <span class="dv">500</span>, <span class="at">sequenceContext =</span> <span class="st">"A"</span>, </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">tracks =</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">trackData =</span> <span class="st">"mod_prob"</span>, <span class="at">trackType =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>                              <span class="at">interpolate =</span> <span class="cn">TRUE</span>),</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">list</span>(<span class="at">trackData =</span> <span class="st">"FracMod"</span>, <span class="at">trackType =</span> <span class="st">"Smooth"</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colorBy =</span> <span class="st">"group"</span>, <span class="at">highlightRegions =</span> reg))) <span class="sc">+</span> </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-top-dmr-scanning" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-top-dmr-scanning-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="scanning_files/figure-html/fig-top-dmr-scanning-1.png" id="fig-top-dmr-scanning" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-top-dmr-scanning-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.3
</figcaption>
</figure>
</div>
</div>
</div>
<section id="choosing-suitable-values-for-thresholds-and-smoothing-parameters" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="choosing-suitable-values-for-thresholds-and-smoothing-parameters"><span class="header-section-number">6.1.1</span> Choosing suitable values for thresholds and smoothing parameters</h3>
<p>In the example above, we smooth the window scores and fuse neighboring windows with smoothed scores above a certain threshold (3). The parameters for this step of the analysis were determined by inspecting the intermediate results obtained from applying the quantification and scoring functions independently to a small region of the genome. Here we illustrate how this can be done, using a 5MB region on chromosome 19. First, we call the <a href="https://fmicompbio.github.io/footprintR/reference/quantifyWindowsInRegion.html"><code>quantifyWindowsInRegion</code></a> function, with <a href="https://fmicompbio.github.io/footprintR/reference/sumNmodNvalid.html"><code>sumNmodNvalid</code></a> as the <em>quantFunction</em>, to get the total and modified nucleotide and the fraction of modified nucleotides for each window. The windows will be automatically defined within the specified <code>region</code>, given the indicated window size and step.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define the small region</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>regSmall <span class="ot">&lt;-</span> <span class="st">"chr19:31000000-36000000"</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define windows in the small region and quantify data in windows</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">quantifyWindowsInRegion</span>(<span class="at">bamfiles =</span> bamfiles,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">region =</span> regSmall,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">modbase =</span> <span class="st">"a"</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">modProbThreshold =</span> <span class="fl">0.5</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sampleAnnot =</span> sampleAnnot,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>                              <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm),</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>,</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sequenceReference =</span> gnm,</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>                              <span class="at">windowMode =</span> <span class="st">"fixed"</span>,</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>                              <span class="at">windowSize =</span> <span class="dv">30</span>,</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>                              <span class="at">windowStep =</span> <span class="dv">15</span>,</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>                              <span class="at">quantFunction =</span> <span class="st">"sumNmodNvalid"</span>,</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>                              <span class="at">BPPARAM =</span> bpParam,</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>                              <span class="at">verbose =</span> <span class="fu">interactive</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The output of <a href="https://fmicompbio.github.io/footprintR/reference/quantifyWindowsInRegion.html"><code>quantifyWindowsInRegion</code></a> is a <code>SummarizedExperiment</code> object, with assays for the <code>Nmod</code>, <code>Nvalid</code> and <code>FracMod</code> values for each window in each sample.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 333073 4 
metadata(1): readLevelData
assays(3): Nmod Nvalid FracMod
rownames(333073): 1 2 ... 333331 333332
rowData names(0):
colnames(4): mESC_wt_6mA_rep1 mESC_wt_6mA_rep2 mESC_TKO_6mA_rep1
  mESC_TKO_6mA_rep2
colData names(3): sample modbase group</code></pre>
</div>
</div>
<p>Next, we apply the score function (<a href="https://fmicompbio.github.io/footprintR/reference/getDifferentiallyModifiedWindows.html"><code>getDifferentiallyModifiedWindows</code></a>) to perform the differential methylation analysis for each window.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">getDifferentiallyModifiedWindows</span>(se, <span class="at">verbose =</span> <span class="fu">interactive</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>This function (as all other score functions compatible with the <a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a> scanning framework) returns a <code>GRanges</code> object containing the windows, with scores as metadata columns.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 333073 ranges and 9 metadata columns:
           seqnames            ranges strand |     logFC    logCPM        LR
              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
       [1]    chr19 31000000-31000029      * |  0.807304  0.629767   2.45710
       [2]    chr19 31000015-31000044      * |  0.742682  0.697488   2.06101
       [3]    chr19 31000030-31000059      * |  0.636205  0.516105   1.54618
       [4]    chr19 31000045-31000074      * |  0.508839  0.762750   0.95991
       [5]    chr19 31000060-31000089      * |  0.648976  0.770806   1.60876
       ...      ...               ...    ... .       ...       ...       ...
  [333069]    chr19 35999905-35999934      * | 0.1183276 0.0717503 0.0623986
  [333070]    chr19 35999920-35999949      * | 0.2911895 0.5566819 0.3401543
  [333071]    chr19 35999935-35999964      * | 0.0709167 0.5667234 0.0202314
  [333072]    chr19 35999950-35999979      * | 0.2903877 0.2983924 0.3527457
  [333073]    chr19 35999965-35999994      * | 0.7348412 0.7163173 1.8874211
              PValue       FDR dirNegLog10PValue FracMod_wt FracMod_TKO
           &lt;numeric&gt; &lt;numeric&gt;         &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
       [1]  0.116994  0.402987          0.931835   0.129820    0.200790
       [2]  0.151110  0.437962          0.820708   0.145305    0.213280
       [3]  0.213700  0.494005          0.670195   0.141810    0.190575
       [4]  0.327210  0.581068          0.485174   0.136349    0.176522
       [5]  0.204666  0.486337          0.688954   0.165851    0.233447
       ...       ...       ...               ...        ...         ...
  [333069]  0.802744  0.891057         0.0954228  0.3054885    0.314199
  [333070]  0.559740  0.739766         0.2520135  0.2349981    0.259468
  [333071]  0.886893  0.939795         0.0521289  0.2306079    0.229048
  [333072]  0.552563  0.735028         0.2576182  0.1846282    0.212466
  [333073]  0.169493  0.455429         0.7708480  0.0793377    0.125558
           DeltaFracMod
              &lt;numeric&gt;
       [1]    0.0709699
       [2]    0.0679755
       [3]    0.0487643
       [4]    0.0401731
       [5]    0.0675964
       ...          ...
  [333069]   0.00871096
  [333070]   0.02446953
  [333071]  -0.00155955
  [333072]   0.02783778
  [333073]   0.04621999
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>At this point, we can either keep the results on the individual window level, or attempt to fuse neighboring windows with high scores, using the <a href="https://fmicompbio.github.io/footprintR/reference/processWindowScores.html"><code>processWindowScores</code></a> function. We will illustrate the latter using the <code>dirNegLog10PValue</code> as the score column. First, we plot the estimates of this score along the genomic region (just a small part of it, for visibility).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(gr[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>]), <span class="fu">aes</span>(<span class="at">x =</span> start, <span class="at">y =</span> dirNegLog10PValue)) <span class="sc">+</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span> </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-raw-scores-dmw" class="quarto-float quarto-figure quarto-figure-center anchored" width="768">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-raw-scores-dmw-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="scanning_files/figure-html/fig-raw-scores-dmw-1.png" id="fig-raw-scores-dmw" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-raw-scores-dmw-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.4
</figcaption>
</figure>
</div>
</div>
</div>
<p>To determine suitable smoothing parameters, we can call the <a href="https://fmicompbio.github.io/footprintR/reference/processWindow.html"><code>processWindow</code></a> with <code>scoreAction = "smooth"</code>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>(grsmooth <span class="ot">&lt;-</span> <span class="fu">processWindowScores</span>(gr, <span class="at">scoreCol =</span> <span class="st">"dirNegLog10PValue"</span>, </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">scoreAction =</span> <span class="st">"smooth"</span>, <span class="at">minperiod =</span> <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 333073 ranges and 9 metadata columns:
           seqnames            ranges strand |     logFC    logCPM        LR
              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
       [1]    chr19 31000000-31000029      * |  0.807304  0.629767   2.45710
       [2]    chr19 31000015-31000044      * |  0.742682  0.697488   2.06101
       [3]    chr19 31000030-31000059      * |  0.636205  0.516105   1.54618
       [4]    chr19 31000045-31000074      * |  0.508839  0.762750   0.95991
       [5]    chr19 31000060-31000089      * |  0.648976  0.770806   1.60876
       ...      ...               ...    ... .       ...       ...       ...
  [333069]    chr19 35999905-35999934      * | 0.1183276 0.0717503 0.0623986
  [333070]    chr19 35999920-35999949      * | 0.2911895 0.5566819 0.3401543
  [333071]    chr19 35999935-35999964      * | 0.0709167 0.5667234 0.0202314
  [333072]    chr19 35999950-35999979      * | 0.2903877 0.2983924 0.3527457
  [333073]    chr19 35999965-35999994      * | 0.7348412 0.7163173 1.8874211
              PValue       FDR dirNegLog10PValue FracMod_wt FracMod_TKO
           &lt;numeric&gt; &lt;numeric&gt;         &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
       [1]  0.116994  0.402987          0.877458   0.129820    0.200790
       [2]  0.151110  0.437962          0.764158   0.145305    0.213280
       [3]  0.213700  0.494005          0.661064   0.141810    0.190575
       [4]  0.327210  0.581068          0.706547   0.136349    0.176522
       [5]  0.204666  0.486337          0.911285   0.165851    0.233447
       ...       ...       ...               ...        ...         ...
  [333069]  0.802744  0.891057         0.0687533  0.3054885    0.314199
  [333070]  0.559740  0.739766         0.1374055  0.2349981    0.259468
  [333071]  0.886893  0.939795         0.2544209  0.2306079    0.229048
  [333072]  0.552563  0.735028         0.4199374  0.1846282    0.212466
  [333073]  0.169493  0.455429         0.5510439  0.0793377    0.125558
           DeltaFracMod
              &lt;numeric&gt;
       [1]    0.0709699
       [2]    0.0679755
       [3]    0.0487643
       [4]    0.0401731
       [5]    0.0675964
       ...          ...
  [333069]   0.00871096
  [333070]   0.02446953
  [333071]  -0.00155955
  [333072]   0.02783778
  [333073]   0.04621999
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>This replaces the <code>dirNegLog10PValue</code> column with a smoothed version, which we can see by plotting it overlaid on the unsmoothed values.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>plotdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">start</span>(gr[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>]), </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">raw =</span> gr<span class="sc">$</span>dirNegLog10PValue[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>],</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">smoothed =</span> grsmooth<span class="sc">$</span>dirNegLog10PValue[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>]</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">names_to =</span> <span class="st">"type"</span>, <span class="at">values_to =</span> <span class="st">"score"</span>, <span class="sc">-</span>position)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plotdf, <span class="fu">aes</span>(<span class="at">x =</span> position, <span class="at">y =</span> score, <span class="at">color =</span> type)) <span class="sc">+</span> </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="at">raw =</span> <span class="st">"black"</span>, <span class="at">smoothed =</span> <span class="st">"red"</span>)) <span class="sc">+</span> </span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-smooth-scores-dmw" class="quarto-float quarto-figure quarto-figure-center anchored" width="768">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-smooth-scores-dmw-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="scanning_files/figure-html/fig-smooth-scores-dmw-1.png" id="fig-smooth-scores-dmw" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-smooth-scores-dmw-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.5
</figcaption>
</figure>
</div>
</div>
</div>
<p>Based on these values, we can now also decide on a suitable value for the <code>threshold</code> parameter, which will be applied to the smoothed values to determine which windows show a strong signal.</p>
</section>
</section>
<section id="genome-wide-estimation-of-cpg-methylation-levels-within-samples" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="genome-wide-estimation-of-cpg-methylation-levels-within-samples"><span class="header-section-number">6.2</span> Genome-wide estimation of CpG methylation levels (within samples)</h2>
<p>In the example above, the scores were calculated over windows of different size. By setting the window size to 1, we can also calculate single nucleotide-level scores. Here, as an example, we illustrate how to estimate the global distribution of methylation levels across all CpGs in the genome. For time reasons, as above we only consider chromosome 19 here.</p>
<p>In this case, we again use <a href="https://fmicompbio.github.io/footprintR/reference/sumNmodNvalid.html"><code>sumNmodNvalid</code></a> as the <code>quantFunction</code>, as it provides an estimate of the fraction modified bases for each window (in this case, since the window size is 1, for each modified base). We use <a href="https://fmicompbio.github.io/footprintR/reference/getRangesWithAssayValues.html"><code>getRangesWithAssayValues</code></a> as the <code>scoreFunction</code>, which will move the calculated values from the <code>FracMod</code> and <code>Nvalid</code> assays to the metadata columns of the nucleotide-level <code>GRanges</code> object. By setting <code>scoreAction</code> to “pass”, we instruct the scanning framework to not process the nucleotide-level scores further, but simply return the <code>GRanges</code> object with the added metadata columns. Finally, by setting the <code>sequenceContext</code> to “NCG”, we make sure to only retain positions where the genomic sequence is CG.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the bam file to use and generate sample annotation table</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>bamfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">wt2 =</span> <span class="st">"data/mESC_wt_5mCG_5hmCG_rep2.bam"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>sampleAnnot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(bamfiles))</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run scanning</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>chrlens <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(gnm)[<span class="st">"chr19"</span>]</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>rescpgL <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(chrlens), <span class="cf">function</span>(i) {</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scanForHighScoringRegions</span>(</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">bamfiles =</span> bamfiles,</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">sampleAnnot =</span> sampleAnnot,</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">chromosomeLengths =</span> chrlens[i],</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">quantFunction =</span> <span class="st">"sumNmodNvalid"</span>,</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">quantFunctionArgs =</span> <span class="fu">list</span>(),</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreFunction =</span> <span class="st">"getRangesWithAssayValues"</span>,</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreFunctionArgs =</span> <span class="fu">list</span>(<span class="at">assayName =</span> <span class="fu">c</span>(<span class="st">"FracMod"</span>, <span class="st">"Nvalid"</span>)),</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">modbase =</span> <span class="st">"m"</span>,</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">modProbThreshold =</span> <span class="fl">0.5</span>,</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">tileSize =</span> <span class="fl">1e6</span>,</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm),</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContextWidth =</span> <span class="dv">3</span>,</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceReference =</span> gnm,</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContext =</span> <span class="st">"NCG"</span>,</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowMode =</span> <span class="st">"fixed"</span>,</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowSize =</span> <span class="dv">1</span>,</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowStep =</span> <span class="dv">1</span>,</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreCol =</span> <span class="fu">paste0</span>(<span class="st">"FracMod."</span>, <span class="fu">names</span>(bamfiles)[<span class="dv">1</span>]),</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreAction =</span> <span class="st">"pass"</span>,</span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>        <span class="at">BPPARAM =</span> bpParam,</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Next, we plot the distribution of modification fractions.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>plotdf <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">unname</span>(rescpgL[[<span class="dv">1</span>]])) <span class="sc">|&gt;</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(seqnames, start, end, width, strand), </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">".value"</span>, <span class="st">"sample"</span>),</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_pattern =</span> <span class="st">'(FracMod|Nvalid)</span><span class="sc">\\</span><span class="st">.(.*)'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">FracModBin =</span> Hmisc<span class="sc">::</span><span class="fu">cut2</span>(FracMod,<span class="at">cuts =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.1</span>)))</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(plotdf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  seqnames   start     end width strand sample FracMod Nvalid FracModBin
  &lt;fct&gt;      &lt;int&gt;   &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;     
1 chr19    3050119 3050119     1 *      wt2       1         4 [0.9,1.0] 
2 chr19    3050120 3050120     1 *      wt2       1         2 [0.9,1.0] 
3 chr19    3050316 3050316     1 *      wt2       0.75      4 [0.7,0.8) 
4 chr19    3050317 3050317     1 *      wt2       1         2 [0.9,1.0] 
5 chr19    3050603 3050603     1 *      wt2       1         6 [0.9,1.0] 
6 chr19    3050604 3050604     1 *      wt2       1         1 [0.9,1.0] </code></pre>
</div>
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plotdf, <span class="fu">aes</span>(<span class="at">x =</span> FracModBin)) <span class="sc">+</span> </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span> </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"All CpGs"</span>, <span class="at">x =</span> <span class="st">"FracMod (binned)"</span>) <span class="sc">+</span> </span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-cpg-methylation-scanning" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-cpg-methylation-scanning-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="scanning_files/figure-html/fig-cpg-methylation-scanning-1.png" id="fig-cpg-methylation-scanning" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-cpg-methylation-scanning-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6.6
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="annotation-of-predefined-windows" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="annotation-of-predefined-windows"><span class="header-section-number">6.3</span> Annotation of predefined windows</h2>
<p>As mentioned above, the individual functions called by <a href="https://fmicompbio.github.io/footprintR/reference/scanForHighScoringRegions.html"><code>scanForHighScoringRegions</code></a> can also be used independently. This can be helpful in different settings - as already shown, it is helpful in order to determine suitable parameters and thresholds for the smoothing and fusing of windows. Another situation when it is useful is when the windows are predefined rather than obtained by tiling the genome into equally sized bins. To illustrate this use case, here we will use the <a href="https://fmicompbio.github.io/footprintR/reference/quantifyWindowsInRegion.html"><code>quantifyWindowsInRegion</code></a> function, together with the <a href="https://fmicompbio.github.io/footprintR/reference/phasingScoreFourier.html"><code>phasingScoreFourier</code></a> <em>quantFunction</em> to annotate a set of CTCF sites with a score quantifying the degree of nucleosome phasing estimated from the average 6mA signal across reads in the region around the motif. This function calculates a phasing score for each window by decomposing the <code>FracMod</code> values into frequency components and extracting the amplitude of the component corresponding to a period consistent with the average nucleosome distance. Note that we set <code>windowMode</code> to <code>"predefined"</code>, and define the desired windows via the <code>windows</code> argument. In addition, we use larger windows than above (760bp), to have enough data to be able to detect periodic patterns.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>bamfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"data/mESC_wt_6mA_rep1.bam"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bamfiles) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.bam"</span>, <span class="st">""</span>, <span class="fu">basename</span>(bamfiles))</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>sampleAnnot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(bamfiles), </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">group =</span> <span class="st">"wt"</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>period <span class="ot">&lt;-</span> <span class="dv">190</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>windowSize <span class="ot">&lt;-</span> <span class="dv">4</span> <span class="sc">*</span> period</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>numCoef <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>(ctcf_sites <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/ctcf_sites_1000.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1000 ranges and 1 metadata column:
         seqnames              ranges strand |     score
            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
     [1]     chr4 106387210-106387228      - |    21.751
     [2]     chr2   29509735-29509753      + |    16.924
     [3]    chr13 110505258-110505276      - |    19.577
     [4]     chr3 121678692-121678710      + |    18.323
     [5]     chr2   29509695-29509713      + |    17.356
     ...      ...                 ...    ... .       ...
   [996]     chr5   35519412-35519430      + |    19.651
   [997]     chr8   23727858-23727876      + |    11.875
   [998]     chr6 112468049-112468067      + |    16.978
   [999]     chr4   44564745-44564763      + |    22.866
  [1000]     chr8 106101422-106101440      + |    10.054
  -------
  seqinfo: 61 sequences (1 circular) from mm39 genome</code></pre>
</div>
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ctcf_sites <span class="ot">&lt;-</span> <span class="fu">resize</span>(ctcf_sites, <span class="at">width =</span> windowSize, <span class="at">fix =</span> <span class="st">"center"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>ctcf_sites <span class="ot">&lt;-</span> <span class="fu">unstrand</span>(ctcf_sites)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(<span class="at">bamfiles =</span> bamfiles, <span class="at">sampleAnnot =</span> sampleAnnot, </span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">regions =</span> ctcf_sites, <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">sequenceReference =</span> gnm, <span class="at">trim =</span> <span class="cn">TRUE</span>, </span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(se)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>ctcf_sites<span class="sc">$</span>phasing_scores <span class="ot">&lt;-</span> <span class="fu">unlist</span>(parallel<span class="sc">::</span><span class="fu">mclapply</span>(</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">seq_along</span>(ctcf_sites),</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(i) {</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>        tmp <span class="ot">&lt;-</span> <span class="fu">phasingScoreFourier</span>(<span class="at">se =</span> <span class="fu">subsetByOverlaps</span>(se, ctcf_sites[i]),</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">gr =</span> ctcf_sites[i], <span class="at">numCoef =</span> numCoef)</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">nrow</span>(tmp) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>            phasingscores <span class="ot">&lt;-</span> <span class="fu">assay</span>(tmp, <span class="st">"phasingScoreAbs"</span>)[<span class="dv">1</span>, ]</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>            phasingscores <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>        <span class="co"># average across samples (in this case only one sample is used)</span></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mean</span>(phasingscores, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">mc.cores =</span> BiocParallel<span class="sc">::</span><span class="fu">bpnworkers</span>(bpParam)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="session-info" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="session-info"><span class="header-section-number">6.4</span> Session info</h2>
<details>
<summary>
<b> Click to view session info </b>
</summary>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="at">info =</span> <span class="st">"packages"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>═ Session info ═══════════════════════════════════════════════════════════════
─ Packages ───────────────────────────────────────────────────────────────────
 package                                      * version   date (UTC) lib source
 abind                                          1.4-8     2024-09-12 [1] CRAN (R 4.5.0)
 backports                                      1.5.0     2024-05-23 [1] CRAN (R 4.5.0)
 base64enc                                      0.1-3     2015-07-28 [1] CRAN (R 4.5.0)
 Biobase                                      * 2.68.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BiocGenerics                                 * 0.54.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BiocIO                                       * 1.18.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BiocParallel                                 * 1.42.1    2025-06-01 [1] Bioconductor 3.21 (R 4.5.0)
 Biostrings                                   * 2.76.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 bitops                                         1.0-9     2024-10-03 [1] CRAN (R 4.5.0)
 BSgenome                                     * 1.76.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BSgenome.Mmusculus.GENCODE.GRCm39.gencodeM34 * 0.1.0     2025-04-17 [1] Bioconductor
 checkmate                                      2.3.2     2024-07-29 [1] CRAN (R 4.5.0)
 cli                                            3.6.5     2025-04-23 [1] CRAN (R 4.5.0)
 cluster                                        2.1.8.1   2025-03-12 [2] CRAN (R 4.5.0)
 codetools                                      0.2-20    2024-03-31 [2] CRAN (R 4.5.0)
 colorspace                                     2.1-1     2024-07-26 [1] CRAN (R 4.5.0)
 crayon                                         1.5.3     2024-06-20 [1] CRAN (R 4.5.0)
 curl                                           6.3.0     2025-06-06 [1] CRAN (R 4.5.0)
 data.table                                     1.17.4    2025-05-26 [1] CRAN (R 4.5.0)
 DelayedArray                                   0.34.1    2025-04-17 [1] Bioconductor 3.21 (R 4.5.0)
 dichromat                                      2.0-0.1   2022-05-02 [1] CRAN (R 4.5.0)
 digest                                         0.6.37    2024-08-19 [1] CRAN (R 4.5.0)
 dplyr                                        * 1.1.4     2023-11-17 [1] CRAN (R 4.5.0)
 edgeR                                          4.6.2     2025-05-07 [1] Bioconductor 3.21 (R 4.5.0)
 evaluate                                       1.0.3     2025-01-10 [1] CRAN (R 4.5.0)
 farver                                         2.1.2     2024-05-13 [1] CRAN (R 4.5.0)
 fastmap                                        1.2.0     2024-05-15 [1] CRAN (R 4.5.0)
 footprintR                                   * 0.3.5     2025-06-04 [1] Github (fmicompbio/footprintR@612f713)
 foreign                                        0.8-90    2025-03-31 [2] CRAN (R 4.5.0)
 Formula                                        1.2-5     2023-02-24 [1] CRAN (R 4.5.0)
 generics                                     * 0.1.4     2025-05-09 [1] CRAN (R 4.5.0)
 GenomeInfoDb                                 * 1.44.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 GenomeInfoDbData                               1.2.14    2025-04-17 [1] Bioconductor
 GenomicAlignments                              1.44.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 GenomicRanges                                * 1.60.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 ggforce                                        0.4.2     2024-02-19 [1] CRAN (R 4.5.0)
 ggplot2                                      * 3.5.2     2025-04-09 [1] CRAN (R 4.5.0)
 glue                                           1.8.0     2024-09-30 [1] CRAN (R 4.5.0)
 gridExtra                                      2.3       2017-09-09 [1] CRAN (R 4.5.0)
 gtable                                         0.3.6     2024-10-25 [1] CRAN (R 4.5.0)
 Hmisc                                        * 5.2-3     2025-03-16 [1] CRAN (R 4.5.0)
 htmlTable                                      2.4.3     2024-07-21 [1] CRAN (R 4.5.0)
 htmltools                                      0.5.8.1   2024-04-04 [1] CRAN (R 4.5.0)
 htmlwidgets                                    1.6.4     2023-12-06 [1] CRAN (R 4.5.0)
 httr                                           1.4.7     2023-08-15 [1] CRAN (R 4.5.0)
 IRanges                                      * 2.42.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 jsonlite                                       2.0.0     2025-03-27 [1] CRAN (R 4.5.0)
 knitr                                          1.50      2025-03-16 [1] CRAN (R 4.5.0)
 labeling                                       0.4.3     2023-08-29 [1] CRAN (R 4.5.0)
 lattice                                        0.22-7    2025-04-02 [1] CRAN (R 4.5.0)
 lifecycle                                      1.0.4     2023-11-07 [1] CRAN (R 4.5.0)
 limma                                          3.64.1    2025-05-25 [1] Bioconductor 3.21 (R 4.5.0)
 locfit                                         1.5-9.12  2025-03-05 [1] CRAN (R 4.5.0)
 magrittr                                       2.0.3     2022-03-30 [1] CRAN (R 4.5.0)
 MASS                                           7.3-65    2025-02-28 [2] CRAN (R 4.5.0)
 Matrix                                         1.7-3     2025-03-11 [2] CRAN (R 4.5.0)
 MatrixGenerics                               * 1.20.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 matrixStats                                  * 1.5.0     2025-01-07 [1] CRAN (R 4.5.0)
 nnet                                           7.3-20    2025-01-01 [2] CRAN (R 4.5.0)
 patchwork                                    * 1.3.0     2024-09-16 [1] CRAN (R 4.5.0)
 pillar                                         1.10.2    2025-04-05 [1] CRAN (R 4.5.0)
 pkgconfig                                      2.0.3     2019-09-22 [1] CRAN (R 4.5.0)
 polyclip                                       1.10-7    2024-07-23 [1] CRAN (R 4.5.0)
 purrr                                          1.0.4     2025-02-05 [1] CRAN (R 4.5.0)
 R6                                             2.6.1     2025-02-15 [1] CRAN (R 4.5.0)
 RColorBrewer                                   1.1-3     2022-04-03 [1] CRAN (R 4.5.0)
 Rcpp                                           1.0.14    2025-01-12 [1] CRAN (R 4.5.0)
 RCurl                                          1.98-1.17 2025-03-22 [1] CRAN (R 4.5.0)
 restfulr                                       0.0.15    2022-06-16 [1] CRAN (R 4.5.0)
 rjson                                          0.2.23    2024-09-16 [1] CRAN (R 4.5.0)
 rlang                                          1.1.6     2025-04-11 [1] CRAN (R 4.5.0)
 rmarkdown                                      2.29      2024-11-04 [1] CRAN (R 4.5.0)
 rpart                                          4.1.24    2025-01-07 [2] CRAN (R 4.5.0)
 Rsamtools                                      2.24.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 rstudioapi                                     0.17.1    2024-10-22 [1] CRAN (R 4.5.0)
 rtracklayer                                  * 1.68.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 S4Arrays                                       1.8.1     2025-06-01 [1] Bioconductor 3.21 (R 4.5.0)
 S4Vectors                                    * 0.46.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 scales                                         1.4.0     2025-04-24 [1] CRAN (R 4.5.0)
 sessioninfo                                    1.2.3     2025-02-05 [1] CRAN (R 4.5.0)
 signal                                         1.8-1     2024-06-26 [1] CRAN (R 4.5.0)
 SparseArray                                    1.8.0     2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 statmod                                        1.5.0     2023-01-06 [1] CRAN (R 4.5.0)
 stringi                                        1.8.7     2025-03-27 [1] CRAN (R 4.5.0)
 stringr                                      * 1.5.1     2023-11-14 [1] CRAN (R 4.5.0)
 SummarizedExperiment                         * 1.38.1    2025-04-30 [1] Bioconductor 3.21 (R 4.5.0)
 tibble                                         3.3.0     2025-06-08 [1] CRAN (R 4.5.0)
 tidyr                                        * 1.3.1     2024-01-24 [1] CRAN (R 4.5.0)
 tidyselect                                     1.2.1     2024-03-11 [1] CRAN (R 4.5.0)
 tweenr                                         2.0.3     2024-02-26 [1] CRAN (R 4.5.0)
 UCSC.utils                                     1.4.0     2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 utf8                                           1.2.6     2025-06-08 [1] CRAN (R 4.5.0)
 vctrs                                          0.6.5     2023-12-01 [1] CRAN (R 4.5.0)
 viridisLite                                    0.4.2     2023-05-02 [1] CRAN (R 4.5.0)
 withr                                          3.0.2     2024-10-28 [1] CRAN (R 4.5.0)
 xfun                                           0.52      2025-04-02 [1] CRAN (R 4.5.0)
 XML                                            3.99-0.18 2025-01-01 [1] CRAN (R 4.5.0)
 XVector                                      * 0.48.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 yaml                                           2.3.10    2024-07-26 [1] CRAN (R 4.5.0)
 zoo                                            1.8-14    2025-04-10 [1] CRAN (R 4.5.0)

 [1] /tungstenfs/groups/gbioinfo/Appz/R/BioC/R-4.5-release-foss-2024.05_BioC-3.21-release-foss-2024.05
 [2] /tachyon/groups/gbioinfo/Appz/easybuild/software/R/4.5.0-foss-2024.05/lib64/R/library
 * ── Packages attached to the search path.

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</details>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/fmicompbio\.github\.io\/footprintR_book");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./nucleosomes.html" class="pagination-link" aria-label="Nucleosome analyses">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Nucleosome analyses</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./grouping-reads.html" class="pagination-link" aria-label="Grouping reads">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Grouping reads</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb30" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Scanning the genome for regions of interest {#sec-scanning}</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">fmicompbio/footprintR</span><span class="co">]</span>{.githubpkg} provides a flexible scanning framework, which can be used to partition the genome (or a subset thereof) into windows and calculate a custom score for each window. </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>Optionally, neighboring windows with similar scores can then be merged into larger signal regions. </span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>In this chapter, we show some examples of how this framework can be used to scan the genome for various types of signals of interest.</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>The figure below illustrates the general idea of the scanning framework.</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/scan-framework-v2.png)</span>{#fig-scan-framework .lightbox fig-alt="Schematic of the footprintR genome scanning framework." height=600}</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>The first step is to define the windows and calculate some quantity of interest for each of them, using the <span class="co">[</span><span class="ot">quantifyWindowsInRegion</span><span class="co">]</span>{.fn} function.</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>Next, these quantities are passed to a _score function_, which summarizes the input data into a score for each window - this can, for example, perform a differential methylation analysis based on the counts of methylated and unmethylated nucleotides obtained from <span class="co">[</span><span class="ot">quantifyWindowsInRegion</span><span class="co">]</span>{.fn}.</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>Finally, the window scores are processed using <span class="co">[</span><span class="ot">processWindowScores</span><span class="co">]</span>{.fn}. </span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>This function allows neighboring windows with similar scores to be merged into larger signal regions. </span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>The user can choose to either call the functions directly, or use the wrapper function <span class="co">[</span><span class="ot">scanForHighScoringRegions</span><span class="co">]</span>{.fn}.</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>Thanks to the modularity, the scanning framework is flexible and can accommodate many different use cases by selecting a suitable quantification function, scoring function and aggregation of the scores. </span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>The table below lists some of the applications that can be accommodated with the functions that are provided within <span class="co">[</span><span class="ot">fmicompbio/footprintR</span><span class="co">]</span>{.githubpkg}. </span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>However, the user can also define their own functions for specific use cases.</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/scan-useCases.png)</span>{#fig-scan-usecases .lightbox fig-alt="Table of use cases that can be addressed using the genome scanning framework in footprintR." height=300}</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>Below we will show several examples of using the scanning framework for different purposes. </span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-packages</span></span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a>BSgenomeName <span class="ot">&lt;-</span> <span class="st">"BSgenome.Mmusculus.GENCODE.GRCm39.gencodeM34"</span></span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(footprintR)</span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenomeName, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Hmisc)</span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb30-43"><a href="#cb30-43" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb30-44"><a href="#cb30-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Load genome</span></span>
<span id="cb30-46"><a href="#cb30-46" aria-hidden="true" tabindex="-1"></a>gnm <span class="ot">&lt;-</span> <span class="fu">get</span>(BSgenomeName)</span>
<span id="cb30-47"><a href="#cb30-47" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(gnm) <span class="ot">&lt;-</span> <span class="st">"mm39"</span></span>
<span id="cb30-48"><a href="#cb30-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-49"><a href="#cb30-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Define parallelization backend</span></span>
<span id="cb30-50"><a href="#cb30-50" aria-hidden="true" tabindex="-1"></a>bpParam <span class="ot">&lt;-</span> <span class="fu">MulticoreParam</span>(<span class="at">workers =</span> <span class="dv">24</span>L, <span class="at">RNGseed =</span> <span class="dv">42</span>L)</span>
<span id="cb30-51"><a href="#cb30-51" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-52"><a href="#cb30-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-53"><a href="#cb30-53" aria-hidden="true" tabindex="-1"></a><span class="fu">## Genome-wide DMR scanning (between samples)</span></span>
<span id="cb30-54"><a href="#cb30-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-55"><a href="#cb30-55" aria-hidden="true" tabindex="-1"></a>The first use case involves scanning for differentially methylated regions (DMRs) between two conditions, based on 6mA data.</span>
<span id="cb30-56"><a href="#cb30-56" aria-hidden="true" tabindex="-1"></a>In our case, we will compare wild-type mouse ES cells to 'TKO' mouse ES cells where the three DNA methyltransferases (Dnmts) have been knocked out.</span>
<span id="cb30-57"><a href="#cb30-57" aria-hidden="true" tabindex="-1"></a>For time reasons, we limit the scanning to chromosome 19 - the most efficient way of performing a genome-wide analysis is typically to parallelize over chromosomes (e.g. by replacing the <span class="in">`lapply`</span> call in the code below by a parallelized version).</span>
<span id="cb30-58"><a href="#cb30-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-59"><a href="#cb30-59" aria-hidden="true" tabindex="-1"></a>We tile the genome into 30bp windows (<span class="in">`windowSize`</span>), shifted successively by 15bp (<span class="in">`windowStep`</span>) to achieve a high resolution and minimize the risk of splitting a region of interest in the middle (thereby potentially losing the signal). </span>
<span id="cb30-60"><a href="#cb30-60" aria-hidden="true" tabindex="-1"></a>We then use the <span class="co">[</span><span class="ot">sumNmodNvalid</span><span class="co">]</span>{.fn} function to count the number of modified As for each window, as well as the total number of As.</span>
<span id="cb30-61"><a href="#cb30-61" aria-hidden="true" tabindex="-1"></a>The classification of nucleotides is obtained by thresholding the modification probability using the <span class="in">`modProbThreshold`</span> value (0.5).</span>
<span id="cb30-62"><a href="#cb30-62" aria-hidden="true" tabindex="-1"></a>Given the modified and total count for each sample, we next use the <span class="co">[</span><span class="ot">getDifferentiallyModifiedWindows</span><span class="co">]</span>{.fn} score function to perform the differential methylation analysis for each window. </span>
<span id="cb30-63"><a href="#cb30-63" aria-hidden="true" tabindex="-1"></a>This will return several statistics, including the <span class="in">`dirNegLog10PValue`</span>, which is obtained by calculating <span class="in">`-log10(p-value)`</span> and multiplying with the sign of the log-fold change. </span>
<span id="cb30-64"><a href="#cb30-64" aria-hidden="true" tabindex="-1"></a>We will use this statistic (<span class="in">`scoreCol`</span>) as the input for the final processing of the windows.</span>
<span id="cb30-65"><a href="#cb30-65" aria-hidden="true" tabindex="-1"></a>Briefly, the statistics will be smoothed over neighboring windows (the degree of smoothing is controlled by the <span class="in">`minperiod`</span> argument) and subsequently thresholded (<span class="in">`thresh`</span> argument).</span>
<span id="cb30-66"><a href="#cb30-66" aria-hidden="true" tabindex="-1"></a>Neighboring windows where the statistic exceeds this threshold will be fused into a larger region of interest, unless they are separated by more than 50 nucleotides (<span class="in">`maxGap`</span>). </span>
<span id="cb30-67"><a href="#cb30-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-70"><a href="#cb30-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-71"><a href="#cb30-71" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: define-bamfiles</span></span>
<span id="cb30-72"><a href="#cb30-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-73"><a href="#cb30-73" aria-hidden="true" tabindex="-1"></a><span class="co"># define bam files to include</span></span>
<span id="cb30-74"><a href="#cb30-74" aria-hidden="true" tabindex="-1"></a>bamfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"data/mESC_wt_6mA_rep1.bam"</span>, <span class="st">"data/mESC_wt_6mA_rep2.bam"</span>,</span>
<span id="cb30-75"><a href="#cb30-75" aria-hidden="true" tabindex="-1"></a>              <span class="st">"data/mESC_TKO_6mA_rep1.bam"</span>, <span class="st">"data/mESC_TKO_6mA_rep2.bam"</span>)</span>
<span id="cb30-76"><a href="#cb30-76" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bamfiles) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.bam"</span>, <span class="st">""</span>, <span class="fu">basename</span>(bamfiles))</span>
<span id="cb30-77"><a href="#cb30-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-78"><a href="#cb30-78" aria-hidden="true" tabindex="-1"></a><span class="co"># define sample annotation table</span></span>
<span id="cb30-79"><a href="#cb30-79" aria-hidden="true" tabindex="-1"></a>sampleAnnot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(bamfiles),</span>
<span id="cb30-80"><a href="#cb30-80" aria-hidden="true" tabindex="-1"></a>                          <span class="at">group =</span> <span class="fu">factor</span>(<span class="fu">str_extract</span>(<span class="fu">names</span>(bamfiles), <span class="st">"wt|TKO"</span>), </span>
<span id="cb30-81"><a href="#cb30-81" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"wt"</span>, <span class="st">"TKO"</span>)))</span>
<span id="cb30-82"><a href="#cb30-82" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-83"><a href="#cb30-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-86"><a href="#cb30-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-87"><a href="#cb30-87" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr isTRUE(exists("params") &amp;&amp; is.list(params) &amp;&amp; isTRUE(params$recompute))</span></span>
<span id="cb30-88"><a href="#cb30-88" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: run-dmr-scanning</span></span>
<span id="cb30-89"><a href="#cb30-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-90"><a href="#cb30-90" aria-hidden="true" tabindex="-1"></a><span class="co"># run scanning of chromosome 19</span></span>
<span id="cb30-91"><a href="#cb30-91" aria-hidden="true" tabindex="-1"></a>chrlens <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(gnm)[<span class="st">"chr19"</span>]</span>
<span id="cb30-92"><a href="#cb30-92" aria-hidden="true" tabindex="-1"></a>dmrL <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(chrlens), <span class="cf">function</span>(i) {</span>
<span id="cb30-93"><a href="#cb30-93" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scanForHighScoringRegions</span>(</span>
<span id="cb30-94"><a href="#cb30-94" aria-hidden="true" tabindex="-1"></a>        <span class="at">bamfiles =</span> bamfiles,</span>
<span id="cb30-95"><a href="#cb30-95" aria-hidden="true" tabindex="-1"></a>        <span class="at">sampleAnnot =</span> sampleAnnot,</span>
<span id="cb30-96"><a href="#cb30-96" aria-hidden="true" tabindex="-1"></a>        <span class="at">chromosomeLengths =</span> chrlens[i],</span>
<span id="cb30-97"><a href="#cb30-97" aria-hidden="true" tabindex="-1"></a>        <span class="at">quantFunction =</span> <span class="st">"sumNmodNvalid"</span>,</span>
<span id="cb30-98"><a href="#cb30-98" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreFunction =</span> <span class="st">"getDifferentiallyModifiedWindows"</span>,</span>
<span id="cb30-99"><a href="#cb30-99" aria-hidden="true" tabindex="-1"></a>        <span class="at">modbase =</span> <span class="st">"a"</span>,</span>
<span id="cb30-100"><a href="#cb30-100" aria-hidden="true" tabindex="-1"></a>        <span class="at">modProbThreshold =</span> <span class="fl">0.5</span>,</span>
<span id="cb30-101"><a href="#cb30-101" aria-hidden="true" tabindex="-1"></a>        <span class="at">tileSize =</span> <span class="fl">2e+06</span>,</span>
<span id="cb30-102"><a href="#cb30-102" aria-hidden="true" tabindex="-1"></a>        <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm),</span>
<span id="cb30-103"><a href="#cb30-103" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>,</span>
<span id="cb30-104"><a href="#cb30-104" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceReference =</span> gnm,</span>
<span id="cb30-105"><a href="#cb30-105" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb30-106"><a href="#cb30-106" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowMode =</span> <span class="st">"fixed"</span>,</span>
<span id="cb30-107"><a href="#cb30-107" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowSize =</span> <span class="dv">30</span>,</span>
<span id="cb30-108"><a href="#cb30-108" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowStep =</span> <span class="dv">15</span>,</span>
<span id="cb30-109"><a href="#cb30-109" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreCol =</span> <span class="st">"dirNegLog10PValue"</span>,</span>
<span id="cb30-110"><a href="#cb30-110" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreAction =</span> <span class="st">"smoothFuse"</span>,</span>
<span id="cb30-111"><a href="#cb30-111" aria-hidden="true" tabindex="-1"></a>        <span class="at">thresh =</span> <span class="dv">3</span>,</span>
<span id="cb30-112"><a href="#cb30-112" aria-hidden="true" tabindex="-1"></a>        <span class="at">minperiod =</span> <span class="dv">3</span>,</span>
<span id="cb30-113"><a href="#cb30-113" aria-hidden="true" tabindex="-1"></a>        <span class="at">maxGap =</span> <span class="dv">50</span>,</span>
<span id="cb30-114"><a href="#cb30-114" aria-hidden="true" tabindex="-1"></a>        <span class="at">BPPARAM =</span> bpParam,</span>
<span id="cb30-115"><a href="#cb30-115" aria-hidden="true" tabindex="-1"></a>        <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb30-116"><a href="#cb30-116" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb30-117"><a href="#cb30-117" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-118"><a href="#cb30-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-121"><a href="#cb30-121" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-122"><a href="#cb30-122" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: !expr isTRUE(exists("params") &amp;&amp; is.list(params) &amp;&amp; isTRUE(params$recompute))</span></span>
<span id="cb30-123"><a href="#cb30-123" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-124"><a href="#cb30-124" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: save-dmr-scanning</span></span>
<span id="cb30-125"><a href="#cb30-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-126"><a href="#cb30-126" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"data/out"</span>)) {</span>
<span id="cb30-127"><a href="#cb30-127" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(<span class="st">"data/out"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-128"><a href="#cb30-128" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-129"><a href="#cb30-129" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dmrL, <span class="st">"data/out/scanning_dmr-list.rds"</span>)</span>
<span id="cb30-130"><a href="#cb30-130" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-131"><a href="#cb30-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-134"><a href="#cb30-134" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-135"><a href="#cb30-135" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-136"><a href="#cb30-136" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-dmr-scanning</span></span>
<span id="cb30-137"><a href="#cb30-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-138"><a href="#cb30-138" aria-hidden="true" tabindex="-1"></a>dmrL <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/out/scanning_dmr-list.rds"</span>)</span>
<span id="cb30-139"><a href="#cb30-139" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-140"><a href="#cb30-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-141"><a href="#cb30-141" aria-hidden="true" tabindex="-1"></a>Each element of the resulting list (in this case, there is only one) contains the inferred DMRs for one chromosome.</span>
<span id="cb30-142"><a href="#cb30-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-145"><a href="#cb30-145" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-146"><a href="#cb30-146" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: explore-dmr-scanning</span></span>
<span id="cb30-147"><a href="#cb30-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-148"><a href="#cb30-148" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(dmrL[[<span class="dv">1</span>]])</span>
<span id="cb30-149"><a href="#cb30-149" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dmrL[[<span class="dv">1</span>]])</span>
<span id="cb30-150"><a href="#cb30-150" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(dmrL[[<span class="dv">1</span>]]<span class="sc">$</span>direction)</span>
<span id="cb30-151"><a href="#cb30-151" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-152"><a href="#cb30-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-153"><a href="#cb30-153" aria-hidden="true" tabindex="-1"></a>We can visualize one of the top hits using <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}.</span>
<span id="cb30-154"><a href="#cb30-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-157"><a href="#cb30-157" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-158"><a href="#cb30-158" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 10</span></span>
<span id="cb30-159"><a href="#cb30-159" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-top-dmr-scanning</span></span>
<span id="cb30-160"><a href="#cb30-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-161"><a href="#cb30-161" aria-hidden="true" tabindex="-1"></a>reg <span class="ot">&lt;-</span> dmrL[[<span class="dv">1</span>]][<span class="fu">which.max</span>(<span class="fu">abs</span>(dmrL[[<span class="dv">1</span>]]<span class="sc">$</span>dirNegLog10PValue))]</span>
<span id="cb30-162"><a href="#cb30-162" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(<span class="at">bamfiles =</span> bamfiles, <span class="at">sampleAnnot =</span> sampleAnnot, </span>
<span id="cb30-163"><a href="#cb30-163" aria-hidden="true" tabindex="-1"></a>                 <span class="at">regions =</span> reg <span class="sc">+</span> <span class="dv">500</span>, <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb30-164"><a href="#cb30-164" aria-hidden="true" tabindex="-1"></a>                 <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb30-165"><a href="#cb30-165" aria-hidden="true" tabindex="-1"></a>                 <span class="at">sequenceReference =</span> gnm, <span class="at">trim =</span> <span class="cn">TRUE</span>, </span>
<span id="cb30-166"><a href="#cb30-166" aria-hidden="true" tabindex="-1"></a>                 <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb30-167"><a href="#cb30-167" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(se)</span>
<span id="cb30-168"><a href="#cb30-168" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(se, <span class="at">region =</span> reg <span class="sc">+</span> <span class="dv">500</span>, <span class="at">sequenceContext =</span> <span class="st">"A"</span>, </span>
<span id="cb30-169"><a href="#cb30-169" aria-hidden="true" tabindex="-1"></a>           <span class="at">tracks =</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">trackData =</span> <span class="st">"mod_prob"</span>, <span class="at">trackType =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb30-170"><a href="#cb30-170" aria-hidden="true" tabindex="-1"></a>                              <span class="at">interpolate =</span> <span class="cn">TRUE</span>),</span>
<span id="cb30-171"><a href="#cb30-171" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">list</span>(<span class="at">trackData =</span> <span class="st">"FracMod"</span>, <span class="at">trackType =</span> <span class="st">"Smooth"</span>,</span>
<span id="cb30-172"><a href="#cb30-172" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colorBy =</span> <span class="st">"group"</span>, <span class="at">highlightRegions =</span> reg))) <span class="sc">+</span> </span>
<span id="cb30-173"><a href="#cb30-173" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>))</span>
<span id="cb30-174"><a href="#cb30-174" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-175"><a href="#cb30-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-176"><a href="#cb30-176" aria-hidden="true" tabindex="-1"></a><span class="fu">### Choosing suitable values for thresholds and smoothing parameters</span></span>
<span id="cb30-177"><a href="#cb30-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-178"><a href="#cb30-178" aria-hidden="true" tabindex="-1"></a>In the example above, we smooth the window scores and fuse neighboring windows with smoothed scores above a certain threshold (3).</span>
<span id="cb30-179"><a href="#cb30-179" aria-hidden="true" tabindex="-1"></a>The parameters for this step of the analysis were determined by inspecting the intermediate results obtained from applying the quantification and scoring functions independently to a small region of the genome.</span>
<span id="cb30-180"><a href="#cb30-180" aria-hidden="true" tabindex="-1"></a>Here we illustrate how this can be done, using a 5MB region on chromosome 19.</span>
<span id="cb30-181"><a href="#cb30-181" aria-hidden="true" tabindex="-1"></a>First, we call the <span class="co">[</span><span class="ot">quantifyWindowsInRegion</span><span class="co">]</span>{.fn} function, with <span class="co">[</span><span class="ot">sumNmodNvalid</span><span class="co">]</span>{.fn} as the _quantFunction_, to get the total and modified nucleotide and the fraction of modified nucleotides for each window.</span>
<span id="cb30-182"><a href="#cb30-182" aria-hidden="true" tabindex="-1"></a>The windows will be automatically defined within the specified <span class="in">`region`</span>, given the indicated window size and step. </span>
<span id="cb30-183"><a href="#cb30-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-186"><a href="#cb30-186" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-187"><a href="#cb30-187" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: quantify-windows-for-dmr</span></span>
<span id="cb30-188"><a href="#cb30-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-189"><a href="#cb30-189" aria-hidden="true" tabindex="-1"></a><span class="co"># define the small region</span></span>
<span id="cb30-190"><a href="#cb30-190" aria-hidden="true" tabindex="-1"></a>regSmall <span class="ot">&lt;-</span> <span class="st">"chr19:31000000-36000000"</span></span>
<span id="cb30-191"><a href="#cb30-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-192"><a href="#cb30-192" aria-hidden="true" tabindex="-1"></a><span class="co"># define windows in the small region and quantify data in windows</span></span>
<span id="cb30-193"><a href="#cb30-193" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">quantifyWindowsInRegion</span>(<span class="at">bamfiles =</span> bamfiles,</span>
<span id="cb30-194"><a href="#cb30-194" aria-hidden="true" tabindex="-1"></a>                              <span class="at">region =</span> regSmall,</span>
<span id="cb30-195"><a href="#cb30-195" aria-hidden="true" tabindex="-1"></a>                              <span class="at">modbase =</span> <span class="st">"a"</span>,</span>
<span id="cb30-196"><a href="#cb30-196" aria-hidden="true" tabindex="-1"></a>                              <span class="at">modProbThreshold =</span> <span class="fl">0.5</span>,</span>
<span id="cb30-197"><a href="#cb30-197" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sampleAnnot =</span> sampleAnnot,</span>
<span id="cb30-198"><a href="#cb30-198" aria-hidden="true" tabindex="-1"></a>                              <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm),</span>
<span id="cb30-199"><a href="#cb30-199" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>,</span>
<span id="cb30-200"><a href="#cb30-200" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sequenceReference =</span> gnm,</span>
<span id="cb30-201"><a href="#cb30-201" aria-hidden="true" tabindex="-1"></a>                              <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb30-202"><a href="#cb30-202" aria-hidden="true" tabindex="-1"></a>                              <span class="at">windowMode =</span> <span class="st">"fixed"</span>,</span>
<span id="cb30-203"><a href="#cb30-203" aria-hidden="true" tabindex="-1"></a>                              <span class="at">windowSize =</span> <span class="dv">30</span>,</span>
<span id="cb30-204"><a href="#cb30-204" aria-hidden="true" tabindex="-1"></a>                              <span class="at">windowStep =</span> <span class="dv">15</span>,</span>
<span id="cb30-205"><a href="#cb30-205" aria-hidden="true" tabindex="-1"></a>                              <span class="at">quantFunction =</span> <span class="st">"sumNmodNvalid"</span>,</span>
<span id="cb30-206"><a href="#cb30-206" aria-hidden="true" tabindex="-1"></a>                              <span class="at">BPPARAM =</span> bpParam,</span>
<span id="cb30-207"><a href="#cb30-207" aria-hidden="true" tabindex="-1"></a>                              <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb30-208"><a href="#cb30-208" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-209"><a href="#cb30-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-210"><a href="#cb30-210" aria-hidden="true" tabindex="-1"></a>The output of <span class="co">[</span><span class="ot">quantifyWindowsInRegion</span><span class="co">]</span>{.fn} is a <span class="in">`SummarizedExperiment`</span> object, with assays for the <span class="in">`Nmod`</span>, <span class="in">`Nvalid`</span> and <span class="in">`FracMod`</span> values for each window in each sample. </span>
<span id="cb30-211"><a href="#cb30-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-214"><a href="#cb30-214" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-215"><a href="#cb30-215" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: show-se</span></span>
<span id="cb30-216"><a href="#cb30-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-217"><a href="#cb30-217" aria-hidden="true" tabindex="-1"></a>se</span>
<span id="cb30-218"><a href="#cb30-218" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-219"><a href="#cb30-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-220"><a href="#cb30-220" aria-hidden="true" tabindex="-1"></a>Next, we apply the score function (<span class="co">[</span><span class="ot">getDifferentiallyModifiedWindows</span><span class="co">]</span>{.fn}) to perform the differential methylation analysis for each window.</span>
<span id="cb30-221"><a href="#cb30-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-224"><a href="#cb30-224" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-225"><a href="#cb30-225" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: get-dmws</span></span>
<span id="cb30-226"><a href="#cb30-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-227"><a href="#cb30-227" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">getDifferentiallyModifiedWindows</span>(se, <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb30-228"><a href="#cb30-228" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-229"><a href="#cb30-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-230"><a href="#cb30-230" aria-hidden="true" tabindex="-1"></a>This function (as all other score functions compatible with the <span class="co">[</span><span class="ot">fmicompbio/footprintR</span><span class="co">]</span>{.githubpkg} scanning framework) returns a <span class="in">`GRanges`</span> object containing the windows, with scores as metadata columns. </span>
<span id="cb30-231"><a href="#cb30-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-234"><a href="#cb30-234" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-235"><a href="#cb30-235" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: show-dmws</span></span>
<span id="cb30-236"><a href="#cb30-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-237"><a href="#cb30-237" aria-hidden="true" tabindex="-1"></a>gr</span>
<span id="cb30-238"><a href="#cb30-238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-239"><a href="#cb30-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-240"><a href="#cb30-240" aria-hidden="true" tabindex="-1"></a>At this point, we can either keep the results on the individual window level, or attempt to fuse neighboring windows with high scores, using the <span class="co">[</span><span class="ot">processWindowScores</span><span class="co">]</span>{.fn} function.</span>
<span id="cb30-241"><a href="#cb30-241" aria-hidden="true" tabindex="-1"></a>We will illustrate the latter using the <span class="in">`dirNegLog10PValue`</span> as the score column. </span>
<span id="cb30-242"><a href="#cb30-242" aria-hidden="true" tabindex="-1"></a>First, we plot the estimates of this score along the genomic region (just a small part of it, for visibility). </span>
<span id="cb30-243"><a href="#cb30-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-246"><a href="#cb30-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-247"><a href="#cb30-247" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 8</span></span>
<span id="cb30-248"><a href="#cb30-248" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 4</span></span>
<span id="cb30-249"><a href="#cb30-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-raw-scores-dmw</span></span>
<span id="cb30-250"><a href="#cb30-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-251"><a href="#cb30-251" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(gr[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>]), <span class="fu">aes</span>(<span class="at">x =</span> start, <span class="at">y =</span> dirNegLog10PValue)) <span class="sc">+</span> </span>
<span id="cb30-252"><a href="#cb30-252" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span> </span>
<span id="cb30-253"><a href="#cb30-253" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb30-254"><a href="#cb30-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-255"><a href="#cb30-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-256"><a href="#cb30-256" aria-hidden="true" tabindex="-1"></a>To determine suitable smoothing parameters, we can call the <span class="co">[</span><span class="ot">processWindow</span><span class="co">]</span>{.fn} with <span class="in">`scoreAction = "smooth"`</span>.</span>
<span id="cb30-257"><a href="#cb30-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-260"><a href="#cb30-260" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-261"><a href="#cb30-261" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: smooth-scores-dmw</span></span>
<span id="cb30-262"><a href="#cb30-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-263"><a href="#cb30-263" aria-hidden="true" tabindex="-1"></a>(grsmooth <span class="ot">&lt;-</span> <span class="fu">processWindowScores</span>(gr, <span class="at">scoreCol =</span> <span class="st">"dirNegLog10PValue"</span>, </span>
<span id="cb30-264"><a href="#cb30-264" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">scoreAction =</span> <span class="st">"smooth"</span>, <span class="at">minperiod =</span> <span class="dv">3</span>))</span>
<span id="cb30-265"><a href="#cb30-265" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-266"><a href="#cb30-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-267"><a href="#cb30-267" aria-hidden="true" tabindex="-1"></a>This replaces the <span class="in">`dirNegLog10PValue`</span> column with a smoothed version, which we can see by plotting it overlaid on the unsmoothed values. </span>
<span id="cb30-268"><a href="#cb30-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-271"><a href="#cb30-271" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-272"><a href="#cb30-272" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 8</span></span>
<span id="cb30-273"><a href="#cb30-273" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 4</span></span>
<span id="cb30-274"><a href="#cb30-274" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-smooth-scores-dmw</span></span>
<span id="cb30-275"><a href="#cb30-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-276"><a href="#cb30-276" aria-hidden="true" tabindex="-1"></a>plotdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb30-277"><a href="#cb30-277" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">start</span>(gr[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>]), </span>
<span id="cb30-278"><a href="#cb30-278" aria-hidden="true" tabindex="-1"></a>    <span class="at">raw =</span> gr<span class="sc">$</span>dirNegLog10PValue[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>],</span>
<span id="cb30-279"><a href="#cb30-279" aria-hidden="true" tabindex="-1"></a>    <span class="at">smoothed =</span> grsmooth<span class="sc">$</span>dirNegLog10PValue[<span class="dv">6700</span><span class="sc">:</span><span class="dv">7200</span>]</span>
<span id="cb30-280"><a href="#cb30-280" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb30-281"><a href="#cb30-281" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">names_to =</span> <span class="st">"type"</span>, <span class="at">values_to =</span> <span class="st">"score"</span>, <span class="sc">-</span>position)</span>
<span id="cb30-282"><a href="#cb30-282" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plotdf, <span class="fu">aes</span>(<span class="at">x =</span> position, <span class="at">y =</span> score, <span class="at">color =</span> type)) <span class="sc">+</span> </span>
<span id="cb30-283"><a href="#cb30-283" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb30-284"><a href="#cb30-284" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="at">raw =</span> <span class="st">"black"</span>, <span class="at">smoothed =</span> <span class="st">"red"</span>)) <span class="sc">+</span> </span>
<span id="cb30-285"><a href="#cb30-285" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb30-286"><a href="#cb30-286" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-287"><a href="#cb30-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-288"><a href="#cb30-288" aria-hidden="true" tabindex="-1"></a>Based on these values, we can now also decide on a suitable value for the <span class="in">`threshold`</span> parameter, which will be applied to the smoothed values to determine which windows show a strong signal. </span>
<span id="cb30-289"><a href="#cb30-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-290"><a href="#cb30-290" aria-hidden="true" tabindex="-1"></a><span class="fu">## Genome-wide estimation of CpG methylation levels (within samples)</span></span>
<span id="cb30-291"><a href="#cb30-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-292"><a href="#cb30-292" aria-hidden="true" tabindex="-1"></a>In the example above, the scores were calculated over windows of different size. </span>
<span id="cb30-293"><a href="#cb30-293" aria-hidden="true" tabindex="-1"></a>By setting the window size to 1, we can also calculate single nucleotide-level scores. </span>
<span id="cb30-294"><a href="#cb30-294" aria-hidden="true" tabindex="-1"></a>Here, as an example, we illustrate how to estimate the global distribution of methylation levels across all CpGs in the genome. </span>
<span id="cb30-295"><a href="#cb30-295" aria-hidden="true" tabindex="-1"></a>For time reasons, as above we only consider chromosome 19 here. </span>
<span id="cb30-296"><a href="#cb30-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-297"><a href="#cb30-297" aria-hidden="true" tabindex="-1"></a>In this case, we again use <span class="co">[</span><span class="ot">sumNmodNvalid</span><span class="co">]</span>{.fn} as the <span class="in">`quantFunction`</span>, as it provides an estimate of the fraction modified bases for each window (in this case, since the window size is 1, for each modified base). </span>
<span id="cb30-298"><a href="#cb30-298" aria-hidden="true" tabindex="-1"></a>We use <span class="co">[</span><span class="ot">getRangesWithAssayValues</span><span class="co">]</span>{.fn} as the <span class="in">`scoreFunction`</span>, which will move the calculated values from the <span class="in">`FracMod`</span> and <span class="in">`Nvalid`</span> assays to the metadata columns of the nucleotide-level <span class="in">`GRanges`</span> object. </span>
<span id="cb30-299"><a href="#cb30-299" aria-hidden="true" tabindex="-1"></a>By setting <span class="in">`scoreAction`</span> to "pass", we instruct the scanning framework to not process the nucleotide-level scores further, but simply return the <span class="in">`GRanges`</span> object with the added metadata columns. </span>
<span id="cb30-300"><a href="#cb30-300" aria-hidden="true" tabindex="-1"></a>Finally, by setting the <span class="in">`sequenceContext`</span> to "NCG", we make sure to only retain positions where the genomic sequence is CG. </span>
<span id="cb30-301"><a href="#cb30-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-302"><a href="#cb30-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-305"><a href="#cb30-305" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-306"><a href="#cb30-306" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: run-cpg-methylation-scanning</span></span>
<span id="cb30-307"><a href="#cb30-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-308"><a href="#cb30-308" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the bam file to use and generate sample annotation table</span></span>
<span id="cb30-309"><a href="#cb30-309" aria-hidden="true" tabindex="-1"></a>bamfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">wt2 =</span> <span class="st">"data/mESC_wt_5mCG_5hmCG_rep2.bam"</span>)</span>
<span id="cb30-310"><a href="#cb30-310" aria-hidden="true" tabindex="-1"></a>sampleAnnot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(bamfiles))</span>
<span id="cb30-311"><a href="#cb30-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-312"><a href="#cb30-312" aria-hidden="true" tabindex="-1"></a><span class="co"># Run scanning</span></span>
<span id="cb30-313"><a href="#cb30-313" aria-hidden="true" tabindex="-1"></a>chrlens <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(gnm)[<span class="st">"chr19"</span>]</span>
<span id="cb30-314"><a href="#cb30-314" aria-hidden="true" tabindex="-1"></a>rescpgL <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(chrlens), <span class="cf">function</span>(i) {</span>
<span id="cb30-315"><a href="#cb30-315" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scanForHighScoringRegions</span>(</span>
<span id="cb30-316"><a href="#cb30-316" aria-hidden="true" tabindex="-1"></a>        <span class="at">bamfiles =</span> bamfiles,</span>
<span id="cb30-317"><a href="#cb30-317" aria-hidden="true" tabindex="-1"></a>        <span class="at">sampleAnnot =</span> sampleAnnot,</span>
<span id="cb30-318"><a href="#cb30-318" aria-hidden="true" tabindex="-1"></a>        <span class="at">chromosomeLengths =</span> chrlens[i],</span>
<span id="cb30-319"><a href="#cb30-319" aria-hidden="true" tabindex="-1"></a>        <span class="at">quantFunction =</span> <span class="st">"sumNmodNvalid"</span>,</span>
<span id="cb30-320"><a href="#cb30-320" aria-hidden="true" tabindex="-1"></a>        <span class="at">quantFunctionArgs =</span> <span class="fu">list</span>(),</span>
<span id="cb30-321"><a href="#cb30-321" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreFunction =</span> <span class="st">"getRangesWithAssayValues"</span>,</span>
<span id="cb30-322"><a href="#cb30-322" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreFunctionArgs =</span> <span class="fu">list</span>(<span class="at">assayName =</span> <span class="fu">c</span>(<span class="st">"FracMod"</span>, <span class="st">"Nvalid"</span>)),</span>
<span id="cb30-323"><a href="#cb30-323" aria-hidden="true" tabindex="-1"></a>        <span class="at">modbase =</span> <span class="st">"m"</span>,</span>
<span id="cb30-324"><a href="#cb30-324" aria-hidden="true" tabindex="-1"></a>        <span class="at">modProbThreshold =</span> <span class="fl">0.5</span>,</span>
<span id="cb30-325"><a href="#cb30-325" aria-hidden="true" tabindex="-1"></a>        <span class="at">tileSize =</span> <span class="fl">1e6</span>,</span>
<span id="cb30-326"><a href="#cb30-326" aria-hidden="true" tabindex="-1"></a>        <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm),</span>
<span id="cb30-327"><a href="#cb30-327" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContextWidth =</span> <span class="dv">3</span>,</span>
<span id="cb30-328"><a href="#cb30-328" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceReference =</span> gnm,</span>
<span id="cb30-329"><a href="#cb30-329" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequenceContext =</span> <span class="st">"NCG"</span>,</span>
<span id="cb30-330"><a href="#cb30-330" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowMode =</span> <span class="st">"fixed"</span>,</span>
<span id="cb30-331"><a href="#cb30-331" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowSize =</span> <span class="dv">1</span>,</span>
<span id="cb30-332"><a href="#cb30-332" aria-hidden="true" tabindex="-1"></a>        <span class="at">windowStep =</span> <span class="dv">1</span>,</span>
<span id="cb30-333"><a href="#cb30-333" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreCol =</span> <span class="fu">paste0</span>(<span class="st">"FracMod."</span>, <span class="fu">names</span>(bamfiles)[<span class="dv">1</span>]),</span>
<span id="cb30-334"><a href="#cb30-334" aria-hidden="true" tabindex="-1"></a>        <span class="at">scoreAction =</span> <span class="st">"pass"</span>,</span>
<span id="cb30-335"><a href="#cb30-335" aria-hidden="true" tabindex="-1"></a>        <span class="at">BPPARAM =</span> bpParam,</span>
<span id="cb30-336"><a href="#cb30-336" aria-hidden="true" tabindex="-1"></a>        <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb30-337"><a href="#cb30-337" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb30-338"><a href="#cb30-338" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb30-339"><a href="#cb30-339" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-340"><a href="#cb30-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-341"><a href="#cb30-341" aria-hidden="true" tabindex="-1"></a>Next, we plot the distribution of modification fractions. </span>
<span id="cb30-342"><a href="#cb30-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-345"><a href="#cb30-345" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-346"><a href="#cb30-346" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-cpg-methylation-scanning</span></span>
<span id="cb30-347"><a href="#cb30-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-348"><a href="#cb30-348" aria-hidden="true" tabindex="-1"></a>plotdf <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">unname</span>(rescpgL[[<span class="dv">1</span>]])) <span class="sc">|&gt;</span></span>
<span id="cb30-349"><a href="#cb30-349" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(seqnames, start, end, width, strand), </span>
<span id="cb30-350"><a href="#cb30-350" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">".value"</span>, <span class="st">"sample"</span>),</span>
<span id="cb30-351"><a href="#cb30-351" aria-hidden="true" tabindex="-1"></a>                        <span class="at">names_pattern =</span> <span class="st">'(FracMod|Nvalid)</span><span class="sc">\\</span><span class="st">.(.*)'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb30-352"><a href="#cb30-352" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">FracModBin =</span> Hmisc<span class="sc">::</span><span class="fu">cut2</span>(FracMod,<span class="at">cuts =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.1</span>)))</span>
<span id="cb30-353"><a href="#cb30-353" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(plotdf)</span>
<span id="cb30-354"><a href="#cb30-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-355"><a href="#cb30-355" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plotdf, <span class="fu">aes</span>(<span class="at">x =</span> FracModBin)) <span class="sc">+</span> </span>
<span id="cb30-356"><a href="#cb30-356" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span> </span>
<span id="cb30-357"><a href="#cb30-357" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"All CpGs"</span>, <span class="at">x =</span> <span class="st">"FracMod (binned)"</span>) <span class="sc">+</span> </span>
<span id="cb30-358"><a href="#cb30-358" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb30-359"><a href="#cb30-359" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb30-360"><a href="#cb30-360" aria-hidden="true" tabindex="-1"></a>          <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb30-361"><a href="#cb30-361" aria-hidden="true" tabindex="-1"></a>          <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>))</span>
<span id="cb30-362"><a href="#cb30-362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-363"><a href="#cb30-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-364"><a href="#cb30-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-365"><a href="#cb30-365" aria-hidden="true" tabindex="-1"></a><span class="fu">## Annotation of predefined windows</span></span>
<span id="cb30-366"><a href="#cb30-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-367"><a href="#cb30-367" aria-hidden="true" tabindex="-1"></a>As mentioned above, the individual functions called by <span class="co">[</span><span class="ot">scanForHighScoringRegions</span><span class="co">]</span>{.fn} can also be used independently. </span>
<span id="cb30-368"><a href="#cb30-368" aria-hidden="true" tabindex="-1"></a>This can be helpful in different settings - as already shown, it is helpful in order to determine suitable parameters and thresholds for the smoothing and fusing of windows. </span>
<span id="cb30-369"><a href="#cb30-369" aria-hidden="true" tabindex="-1"></a>Another situation when it is useful is when the windows are predefined rather than obtained by tiling the genome into equally sized bins. </span>
<span id="cb30-370"><a href="#cb30-370" aria-hidden="true" tabindex="-1"></a>To illustrate this use case, here we will use the <span class="co">[</span><span class="ot">quantifyWindowsInRegion</span><span class="co">]</span>{.fn} function, together with the <span class="co">[</span><span class="ot">phasingScoreFourier</span><span class="co">]</span>{.fn} _quantFunction_ to annotate a set of CTCF sites with a score quantifying the degree of nucleosome phasing estimated from the average 6mA signal across reads in the region around the motif.</span>
<span id="cb30-371"><a href="#cb30-371" aria-hidden="true" tabindex="-1"></a>This function calculates a phasing score for each window by decomposing the <span class="in">`FracMod`</span> values into frequency components and extracting the amplitude of the component corresponding to a period consistent with the average nucleosome distance.</span>
<span id="cb30-372"><a href="#cb30-372" aria-hidden="true" tabindex="-1"></a>Note that we set <span class="in">`windowMode`</span> to <span class="in">`"predefined"`</span>, and define the desired windows via the <span class="in">`windows`</span> argument. </span>
<span id="cb30-373"><a href="#cb30-373" aria-hidden="true" tabindex="-1"></a>In addition, we use larger windows than above (760bp), to have enough data to be able to detect periodic patterns. </span>
<span id="cb30-374"><a href="#cb30-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-377"><a href="#cb30-377" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-378"><a href="#cb30-378" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: add-phasing-scores</span></span>
<span id="cb30-379"><a href="#cb30-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-380"><a href="#cb30-380" aria-hidden="true" tabindex="-1"></a>bamfiles <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"data/mESC_wt_6mA_rep1.bam"</span>)</span>
<span id="cb30-381"><a href="#cb30-381" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bamfiles) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.bam"</span>, <span class="st">""</span>, <span class="fu">basename</span>(bamfiles))</span>
<span id="cb30-382"><a href="#cb30-382" aria-hidden="true" tabindex="-1"></a>sampleAnnot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(bamfiles), </span>
<span id="cb30-383"><a href="#cb30-383" aria-hidden="true" tabindex="-1"></a>                          <span class="at">group =</span> <span class="st">"wt"</span>)</span>
<span id="cb30-384"><a href="#cb30-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-385"><a href="#cb30-385" aria-hidden="true" tabindex="-1"></a>period <span class="ot">&lt;-</span> <span class="dv">190</span></span>
<span id="cb30-386"><a href="#cb30-386" aria-hidden="true" tabindex="-1"></a>windowSize <span class="ot">&lt;-</span> <span class="dv">4</span> <span class="sc">*</span> period</span>
<span id="cb30-387"><a href="#cb30-387" aria-hidden="true" tabindex="-1"></a>numCoef <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb30-388"><a href="#cb30-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-389"><a href="#cb30-389" aria-hidden="true" tabindex="-1"></a>(ctcf_sites <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/ctcf_sites_1000.rds"</span>))</span>
<span id="cb30-390"><a href="#cb30-390" aria-hidden="true" tabindex="-1"></a>ctcf_sites <span class="ot">&lt;-</span> <span class="fu">resize</span>(ctcf_sites, <span class="at">width =</span> windowSize, <span class="at">fix =</span> <span class="st">"center"</span>)</span>
<span id="cb30-391"><a href="#cb30-391" aria-hidden="true" tabindex="-1"></a>ctcf_sites <span class="ot">&lt;-</span> <span class="fu">unstrand</span>(ctcf_sites)</span>
<span id="cb30-392"><a href="#cb30-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-393"><a href="#cb30-393" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(<span class="at">bamfiles =</span> bamfiles, <span class="at">sampleAnnot =</span> sampleAnnot, </span>
<span id="cb30-394"><a href="#cb30-394" aria-hidden="true" tabindex="-1"></a>                 <span class="at">regions =</span> ctcf_sites, <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb30-395"><a href="#cb30-395" aria-hidden="true" tabindex="-1"></a>                 <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb30-396"><a href="#cb30-396" aria-hidden="true" tabindex="-1"></a>                 <span class="at">sequenceReference =</span> gnm, <span class="at">trim =</span> <span class="cn">TRUE</span>, </span>
<span id="cb30-397"><a href="#cb30-397" aria-hidden="true" tabindex="-1"></a>                 <span class="at">verbose =</span> <span class="fu">interactive</span>())</span>
<span id="cb30-398"><a href="#cb30-398" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(se)</span>
<span id="cb30-399"><a href="#cb30-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-400"><a href="#cb30-400" aria-hidden="true" tabindex="-1"></a>ctcf_sites<span class="sc">$</span>phasing_scores <span class="ot">&lt;-</span> <span class="fu">unlist</span>(parallel<span class="sc">::</span><span class="fu">mclapply</span>(</span>
<span id="cb30-401"><a href="#cb30-401" aria-hidden="true" tabindex="-1"></a>    <span class="fu">seq_along</span>(ctcf_sites),</span>
<span id="cb30-402"><a href="#cb30-402" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(i) {</span>
<span id="cb30-403"><a href="#cb30-403" aria-hidden="true" tabindex="-1"></a>        tmp <span class="ot">&lt;-</span> <span class="fu">phasingScoreFourier</span>(<span class="at">se =</span> <span class="fu">subsetByOverlaps</span>(se, ctcf_sites[i]),</span>
<span id="cb30-404"><a href="#cb30-404" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">gr =</span> ctcf_sites[i], <span class="at">numCoef =</span> numCoef)</span>
<span id="cb30-405"><a href="#cb30-405" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">nrow</span>(tmp) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb30-406"><a href="#cb30-406" aria-hidden="true" tabindex="-1"></a>            phasingscores <span class="ot">&lt;-</span> <span class="fu">assay</span>(tmp, <span class="st">"phasingScoreAbs"</span>)[<span class="dv">1</span>, ]</span>
<span id="cb30-407"><a href="#cb30-407" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb30-408"><a href="#cb30-408" aria-hidden="true" tabindex="-1"></a>            phasingscores <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb30-409"><a href="#cb30-409" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb30-410"><a href="#cb30-410" aria-hidden="true" tabindex="-1"></a>        <span class="co"># average across samples (in this case only one sample is used)</span></span>
<span id="cb30-411"><a href="#cb30-411" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mean</span>(phasingscores, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-412"><a href="#cb30-412" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">mc.cores =</span> BiocParallel<span class="sc">::</span><span class="fu">bpnworkers</span>(bpParam)))</span>
<span id="cb30-413"><a href="#cb30-413" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-414"><a href="#cb30-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-415"><a href="#cb30-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-416"><a href="#cb30-416" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session info</span></span>
<span id="cb30-417"><a href="#cb30-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-418"><a href="#cb30-418" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb30-419"><a href="#cb30-419" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;&lt;b&gt;</span>
<span id="cb30-420"><a href="#cb30-420" aria-hidden="true" tabindex="-1"></a>Click to view session info</span>
<span id="cb30-421"><a href="#cb30-421" aria-hidden="true" tabindex="-1"></a>&lt;/b&gt;&lt;/summary&gt;</span>
<span id="cb30-424"><a href="#cb30-424" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-425"><a href="#cb30-425" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: session-info</span></span>
<span id="cb30-426"><a href="#cb30-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-427"><a href="#cb30-427" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="at">info =</span> <span class="st">"packages"</span>)</span>
<span id="cb30-428"><a href="#cb30-428" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-429"><a href="#cb30-429" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb30-430"><a href="#cb30-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-431"><a href="#cb30-431" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/fmicompbio/footprintR_book/edit/main/scanning.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/fmicompbio/footprintR_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer><script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>