<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>4&nbsp; Visualizing regions with plotRegion – **[fmicompbio/footprintR]{.githubpkg}** book</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./nucleosomes.html" rel="next">
<link href="./qc-filtering.html" rel="prev">
<link href="./images/footprintR_logo.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-6082b3b653f69b4bc1378a5897c1dc6c.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./visualize-regions.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualizing regions with <code>plotRegion</code></span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./"><strong></strong></a><strong><a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a></strong> book 
        <div class="sidebar-tools-main">
    <a href="https://github.com/fmicompbio/footprintR_book/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./reading-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Reading data from files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qc-filtering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality control and filtering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visualize-regions.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualizing regions with <code>plotRegion</code></span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./nucleosomes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Nucleosome analyses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scanning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Scanning the genome for regions of interest</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./grouping-reads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Grouping reads</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./multiple-modalities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Combining and comparing multiple data modalities</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#preparation" id="toc-preparation" class="nav-link active" data-scroll-target="#preparation"><span class="header-section-number">4.1</span> Preparation</a></li>
  <li><a href="#read-data" id="toc-read-data" class="nav-link" data-scroll-target="#read-data"><span class="header-section-number">4.2</span> Read data</a></li>
  <li><a href="#read-level-plots" id="toc-read-level-plots" class="nav-link" data-scroll-target="#read-level-plots"><span class="header-section-number">4.3</span> Read-level plots</a></li>
  <li><a href="#summary-level-plots" id="toc-summary-level-plots" class="nav-link" data-scroll-target="#summary-level-plots"><span class="header-section-number">4.4</span> Summary-level plots</a></li>
  <li><a href="#highlight-regions" id="toc-highlight-regions" class="nav-link" data-scroll-target="#highlight-regions"><span class="header-section-number">4.5</span> Highlight regions</a></li>
  <li><a href="#annotation-tracks" id="toc-annotation-tracks" class="nav-link" data-scroll-target="#annotation-tracks"><span class="header-section-number">4.6</span> Annotation tracks</a></li>
  <li><a href="#bigwig-tracks" id="toc-bigwig-tracks" class="nav-link" data-scroll-target="#bigwig-tracks"><span class="header-section-number">4.7</span> <code>BigWig</code> tracks</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="header-section-number">4.8</span> Session info</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/fmicompbio/footprintR_book/edit/main/visualize-regions.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/fmicompbio/footprintR_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-visualize-regions" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualizing regions with <code>plotRegion</code></span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>The <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> function in <a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a> provides a flexible interface for plotting data in a given genomic region. As will be illustrated below, both read-level data, summary-level data (aggregating across all reads overlapping a position) and data from BigWig files can visualized. In addition, predefined regions can be highlighted with background color in any track, and gene models or any other genomic region indicators can be added.</p>
<p>Generally, <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> builds up a plot as a combination of individual <em>tracks</em>, of different <em>types</em> (representing the different data types mentioned above). To add a track to a plot, at least the corresponding <code>trackType</code> and <code>trackData</code> arguments must be specified. A list of the supported track types is available in the documentation of <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> (see the <code>tracks</code> argument). The order and content of the tracks are fully customizable by the user.</p>
<section id="preparation" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="preparation"><span class="header-section-number">4.1</span> Preparation</h2>
<p>We start by loading the required packages. In addition to the software package, we load a <code>BSgenome</code> and <code>TxDb</code> object providing the mouse genome sequence and transcript annotation.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>BSgenomeName <span class="ot">&lt;-</span> <span class="st">"BSgenome.Mmusculus.GENCODE.GRCm39.gencodeM34"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>TxDbName <span class="ot">&lt;-</span> <span class="st">"TxDb.Mmusculus.GENCODE.GRCm39.gencodeM34"</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(footprintR)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenomeName, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TxDbName, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Load genome</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>gnm <span class="ot">&lt;-</span> <span class="fu">get</span>(BSgenomeName)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(gnm) <span class="ot">&lt;-</span> <span class="st">"mm39"</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Load TxDb and create GRangesList with transcript annotations</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> <span class="fu">get</span>(TxDbName)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>ebt <span class="ot">&lt;-</span> <span class="fu">exonsBy</span>(txdb, <span class="st">"tx"</span>, <span class="at">use.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="read-data" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="read-data"><span class="header-section-number">4.2</span> Read data</h2>
<p>To exemplify the plotting capabilities of <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>, we use <code>modBam</code> files from two wild-type samples, for which 6mA modification calling has been performed. We use the <a href="https://fmicompbio.github.io/footprintR/reference/readModBam.html"><code>readModBam</code></a> function to read data from an 800-bp region on chromosome 8 and store it in a <code>SummarizedExperiment</code> object. For more information about reading data with <a href="https://github.com/fmicompbio/footprintR"><em>footprintR</em></a>, see <a href="reading-data.html" class="quarto-xref"><span>Chapter 2</span></a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">bamfiles =</span> <span class="fu">c</span>(<span class="at">wt1 =</span> <span class="st">"data/mESC_wt_6mA_rep1.bam"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">wt2 =</span> <span class="st">"data/mESC_wt_6mA_rep2.bam"</span>),</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">regions =</span> <span class="st">"chr8:39286301-39287100"</span>, </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceReference =</span> gnm</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="read-level-plots" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="read-level-plots"><span class="header-section-number">4.3</span> Read-level plots</h2>
<p>Read-level plots can be generated using either the <code>Heatmap</code> or the <code>Lollipop</code> track type, assuming that there is at least one assay with read-level data present in the provided <code>SummarizedExperiment</code> object. The heatmap visualizes each base with a modification call as a filled rectangle, with optional interpolation between observations. The lollipop plot visualizes each base with a modification call as a filled circle. In both cases, the fill color represents the recorded modification probability of the base. For the read-level plot tracks, the <code>trackData</code> argument specifies the name of the (read-level) assay that should be displayed.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>, </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>),</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-1.png" id="fig-read-level" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.1
</figcaption>
</figure>
</div>
</div>
</div>
<p>In the current plot, reads on both strands are visualized, and can be distinguished based on the position of the modified bases (the As). It is possible to visualize only the reads on one strand, by providing a stranded region to <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800:+"</span>, </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>),</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-stranded" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-stranded-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-stranded-1.png" id="fig-read-level-stranded" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-stranded-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.2
</figcaption>
</figure>
</div>
</div>
</div>
<p>Moreover, we can limit the display to only positions where the annotated genome has a given sequence (here, an A). This is often helpful to minimize the impact of base calling errors.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800:+"</span>, </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-seqcontext" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-seqcontext-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-seqcontext-1.png" id="fig-read-level-seqcontext" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-seqcontext-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.3
</figcaption>
</figure>
</div>
</div>
</div>
<p>Each plot function comes with a set of arguments that can be used to control the way that data is displayed in the track (see the documentation of <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> for the full list). For example, we can add interpolation to our heatmap, disable the automatic clustering of the reads, and add custom track and legend titles. For the lollipop plot, we can similarly change the size of the circles and the linewidth of the circle outline.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-interpolate" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-interpolate-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-interpolate-1.png" id="fig-read-level-interpolate" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-interpolate-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.4
</figcaption>
</figure>
</div>
</div>
</div>
<p>Additional customization options for read-level plots include the possibility to change the color scheme, set the order of the reads, and restricting the set of visualized reads to those that cover a certain fraction of the plot region. Below, we use a ‘inferno’ color scale for the heatmap, and a gradient from white to blue for the lollipop plot. In addition, we order the reads by the average modification fraction across the region (<code>orderReads = "regionAvg"</code>), and exclude any reads that don’t cover at least 95% of the plotted region (<code>minCoveredFraction = 0.95</code>).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-colors" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-colors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-colors-1.png" id="fig-read-level-colors" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-colors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.5
</figcaption>
</figure>
</div>
</div>
</div>
<p>In the plot above, the reads were ordered by the average signal across the entire plotted region. It is also possible to order (or cluster) reads based on the signal in a smaller region, by specifying the <code>orderRegion</code> argument. As an example, we order the reads by the average signal across only the last 20 nucleotides of the plotted region.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-colors-orderregion" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-colors-orderregion-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-colors-orderregion-1.png" id="fig-read-level-colors-orderregion" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-colors-orderregion-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.6
</figcaption>
</figure>
</div>
</div>
</div>
<p>For any ordering, we can also reverse the display order of the reads by setting <code>orderReverse = TRUE</code>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">TRUE</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">TRUE</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-colors-orderregion-orderreverse" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-colors-orderregion-orderreverse-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-colors-orderregion-orderreverse-1.png" id="fig-read-level-colors-orderregion-orderreverse" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-colors-orderregion-orderreverse-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.7
</figcaption>
</figure>
</div>
</div>
</div>
<p>The read-level plots will display all reads overlapping the displayed region. In cases where there are too many reads to display, we recommend subsetting the <code>SummarizedExperiment</code> object before calling <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>. This can be done conveniently using <a href="https://fmicompbio.github.io/footprintR/reference/subsetReads.html"><code>subsetReads</code></a>, which allows you to subset the object keeping either a pre-specified set of reads or a random subset. Here, we illustrate this workflow by subsetting the object to five randomly selected reads per sample before calling <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First subset the object to the plotted region and remove reads not </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># overlapping that region, to restrict the random sampling to reads </span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># overlapping the plot region</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>sesub <span class="ot">&lt;-</span> <span class="fu">filterPositions</span>(se, <span class="at">regions =</span> <span class="st">"chr8:39286600-39286800"</span>, <span class="at">filters =</span> <span class="st">"regions"</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample five random reads for each sample</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>L)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>sesub <span class="ot">&lt;-</span> <span class="fu">subsetReads</span>(sesub, <span class="at">randomSubset =</span> <span class="dv">5</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    sesub, </span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">FALSE</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">FALSE</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-colors-orderregion-randomsubset" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-colors-orderregion-randomsubset-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-colors-orderregion-randomsubset-1.png" id="fig-read-level-colors-orderregion-randomsubset" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-colors-orderregion-randomsubset-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.8
</figcaption>
</figure>
</div>
</div>
</div>
<p>By default, read-level tracks are facetted by sample (the columns of <code>se</code>). We can facet by an arbitrary column of <code>colData(se)</code> by changing the value of the <code>facetBy</code> argument for the track, or set the argument to <code>NULL</code> to disable the facetting and show reads from all samples together.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">facetBy =</span> <span class="cn">NULL</span>),</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>,</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">facetBy =</span> <span class="st">"modbase"</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-nofacet" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-nofacet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-nofacet-1.png" id="fig-read-level-nofacet" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-nofacet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.9
</figcaption>
</figure>
</div>
</div>
</div>
<p>By default, the x-axis of the plots represent genomic coordinates. In some cases (especially if the modified bases are unevenly distributed), the plot may be easier to read if only the modified bases are displayed, evenly spaced along the x-axis. This can be achieved by setting <code>modbaseSpace = TRUE</code> in <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">FALSE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">modbaseSpace =</span> <span class="cn">TRUE</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-modbasespace" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-modbasespace-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-modbasespace-1.png" id="fig-read-level-modbasespace" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-modbasespace-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.10
</figcaption>
</figure>
</div>
</div>
</div>
<p>Another alternative is to anchor the x-axis at a given genomic position, and display the coordinates as distances from this anchor point. This is achieved by setting the <code>referenceCoordinate</code> argument to the numeric position that should be used as the anchor. Note how the x-axis title changes to reflect the current coordinate system.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">referenceCoordinate =</span> <span class="dv">39286700</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-read-level-referencecoordinate" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-referencecoordinate-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-referencecoordinate-1.png" id="fig-read-level-referencecoordinate" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-referencecoordinate-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.11
</figcaption>
</figure>
</div>
</div>
</div>
<p>As <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> returns a <code>patchwork</code>/<code>ggplot2</code> object, even finer control over the plot components is possible. For example, you may want to change the number of ticks on the x axis, which is controlled by <code>ggplot2</code>’s <code>scale_x_continuous()</code>. Changing the x axis scale will throw a warning because by adding the <code>scale_x_continuous()</code>, you are replacing the existing default x axis scale.</p>
<p>Note that if you are plotting multiple tracks with <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>, you need to add <code>scale_x_continuous()</code> using the <code>*</code> or <code>&amp;</code> operator from the <a href="https://cran.r-project.org/package=patchwork"><em>patchwork</em></a> package, instead of <a href="https://cran.r-project.org/package=ggplot2"><em>ggplot2</em></a>’s usual <code>+</code> operator. This will make sure that the plot element (here the new x axis scale) is added to all tracks instead of just the last one:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">referenceCoordinate =</span> <span class="dv">39286700</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>) <span class="sc">*</span> <span class="fu">scale_x_continuous</span>(<span class="at">n.breaks =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for x is already present.
Adding another scale for x, which will replace the existing scale.
Scale for x is already present.
Adding another scale for x, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-read-level-xbreaks" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-read-level-xbreaks-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-read-level-xbreaks-1.png" id="fig-read-level-xbreaks" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-read-level-xbreaks-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.12
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="summary-level-plots" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="summary-level-plots"><span class="header-section-number">4.4</span> Summary-level plots</h2>
<p>In addition to read-level plots, <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> can also display summary-level data (a single value per genomic position and sample), if they are present in the <code>SummarizedExperiment</code> object. These values can be displayed as individual points, and/or with a smooth curve. Summary-level assays can be generated from read-level values using, e.g., the <a href="https://fmicompbio.github.io/footprintR/reference/flattenReadLevelAssay.html"><code>flattenReadLevelAssay</code></a> function.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Flatten read-level assay</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">assayName =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="fu">c</span>(<span class="st">"Nmod"</span>, <span class="st">"Nvalid"</span>, <span class="st">"FracMod"</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Add summary-level track</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39286300-39287100"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Smooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>)</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-summary-level" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-summary-level-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-summary-level-1.png" id="fig-summary-level" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-summary-level-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.13
</figcaption>
</figure>
</div>
</div>
</div>
<p>In an analogous way to the read-level plots, we can control the display of summary information using the arguments to <code>PlotSummaryPointSmooth</code> (see documentation for <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>). For example, we can change the smoothing function and the degree of smoothing. We can also add the individual points (summary values for individual positions) to the summary plot.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add summary-level plot</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39286300-39287100"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>))</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-summary-level-points" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-summary-level-points-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-summary-level-points-1.png" id="fig-summary-level-points" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-summary-level-points-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.14
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="highlight-regions" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="highlight-regions"><span class="header-section-number">4.5</span> Highlight regions</h2>
<p>Sometimes it is helpful to be able to highlight specific genomic regions in the plots. These could be, for example, regions that have been identified as harboring an interesting signal. In <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a>, this can be achieved via the <code>highlightRegions</code> argument to the respective tracks. The input provided to this argument should be a <code>GRanges</code> object.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define regions to highlight</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>reg_to_highlight <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="st">"chr8"</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ranges =</span> <span class="fu">IRanges</span>(</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">start =</span> <span class="fu">c</span>(<span class="dv">39286800</span>, <span class="dv">39286950</span>),</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">end =</span> <span class="fu">c</span>(<span class="dv">39286900</span>, <span class="dv">39287000</span>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39286300-39287100"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>),</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">highlightRegions =</span> reg_to_highlight)</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-highlightregions" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-highlightregions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-highlightregions-1.png" id="fig-highlightregions" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-highlightregions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.15
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="annotation-tracks" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="annotation-tracks"><span class="header-section-number">4.6</span> Annotation tracks</h2>
<p>To further relate the observed data to existing annotations, <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> also allows the user to add one or several tracks representing genomic regions (e.g., gene models). Here, we exemplify how to generate such a track to visualize annotated transcripts in the displayed region.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data from a different region than above (as the previous one does not </span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># overlap annotated genes)</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">bamfiles =</span> <span class="fu">c</span>(<span class="at">wt1 =</span> <span class="st">"data/mESC_wt_6mA_rep1.bam"</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">wt2 =</span> <span class="st">"data/mESC_wt_6mA_rep2.bam"</span>),</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">regions =</span> <span class="st">"chr8:39126917-39128220"</span>, </span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceReference =</span> gnm</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">assayName =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="fu">c</span>(<span class="st">"Nmod"</span>, <span class="st">"Nvalid"</span>, <span class="st">"FracMod"</span>)</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"GenomicRegion"</span>, </span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackData =</span> <span class="fu">subsetByOverlaps</span>(ebt, <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>))),</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>))</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-annotation" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-annotation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-annotation-1.png" id="fig-annotation" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-annotation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.16
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="bigwig-tracks" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="bigwig-tracks"><span class="header-section-number">4.7</span> <code>BigWig</code> tracks</h2>
<p>Finally, <a href="https://fmicompbio.github.io/footprintR/reference/plotRegion.html"><code>plotRegion</code></a> allows the user to include tracks defined by arbitrary BigWig files. Here, we illustrate this by adding a track containing the coverage obtained from a CTCF ChIP-seq experiment. We also modify the colors used to display the summary-level data.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"GenomicRegion"</span>, </span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackData =</span> <span class="fu">subsetByOverlaps</span>(ebt, <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>))),</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"BigWig"</span>, <span class="at">trackData =</span> <span class="fu">c</span>(<span class="at">wt =</span> <span class="st">"data/mESC_wt_CTCF_ChIP_rep1.bw"</span>),</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">yAxisLabel =</span> <span class="st">"CTCF"</span>, <span class="at">showLegend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">colors =</span> <span class="fu">c</span>(<span class="at">wt =</span> <span class="st">"forestgreen"</span>)),</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>),</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>             <span class="at">colors =</span> <span class="fu">c</span>(<span class="at">wt1 =</span> <span class="st">"steelblue"</span>, <span class="at">wt2 =</span> <span class="st">"firebrick2"</span>))</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="fl">0.5</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-bigwig" class="quarto-float quarto-figure quarto-figure-center anchored" width="672">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-bigwig-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="visualize-regions_files/figure-html/fig-bigwig-1.png" id="fig-bigwig" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-bigwig-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4.17
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="session-info" class="level2" data-number="4.8">
<h2 data-number="4.8" class="anchored" data-anchor-id="session-info"><span class="header-section-number">4.8</span> Session info</h2>
<details>
<summary>
<b> Click to view session info </b>
</summary>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="at">info =</span> <span class="st">"packages"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>═ Session info ═══════════════════════════════════════════════════════════════
─ Packages ───────────────────────────────────────────────────────────────────
 package                                      * version   date (UTC) lib source
 abind                                          1.4-8     2024-09-12 [1] CRAN (R 4.5.0)
 AnnotationDbi                                * 1.70.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 Biobase                                      * 2.68.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BiocGenerics                                 * 0.54.1    2025-10-12 [1] Bioconductor 3.21 (R 4.5.0)
 BiocIO                                       * 1.18.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BiocParallel                                 * 1.42.2    2025-09-14 [1] Bioconductor 3.21 (R 4.5.0)
 Biostrings                                   * 2.76.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 bit                                            4.6.0     2025-03-06 [1] CRAN (R 4.5.0)
 bit64                                          4.6.0-1   2025-01-16 [1] CRAN (R 4.5.0)
 bitops                                         1.0-9     2024-10-03 [1] CRAN (R 4.5.0)
 blob                                           1.2.4     2023-03-17 [1] CRAN (R 4.5.0)
 BSgenome                                     * 1.76.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 BSgenome.Mmusculus.GENCODE.GRCm39.gencodeM34 * 0.1.0     2025-04-17 [1] Bioconductor
 cachem                                         1.1.0     2024-05-16 [1] CRAN (R 4.5.0)
 cli                                            3.6.5     2025-04-23 [1] CRAN (R 4.5.0)
 codetools                                      0.2-20    2024-03-31 [2] CRAN (R 4.5.1)
 crayon                                         1.5.3     2024-06-20 [1] CRAN (R 4.5.0)
 curl                                           7.0.0     2025-08-19 [1] CRAN (R 4.5.1)
 data.table                                     1.17.8    2025-07-10 [1] CRAN (R 4.5.0)
 DBI                                            1.2.3     2024-06-02 [1] CRAN (R 4.5.0)
 DelayedArray                                   0.34.1    2025-04-17 [1] Bioconductor 3.21 (R 4.5.0)
 dichromat                                      2.0-0.1   2022-05-02 [1] CRAN (R 4.5.0)
 digest                                         0.6.37    2024-08-19 [1] CRAN (R 4.5.0)
 dplyr                                          1.1.4     2023-11-17 [1] CRAN (R 4.5.0)
 evaluate                                       1.0.5     2025-08-27 [1] CRAN (R 4.5.0)
 farver                                         2.1.2     2024-05-13 [1] CRAN (R 4.5.0)
 fastmap                                        1.2.0     2024-05-15 [1] CRAN (R 4.5.0)
 footprintR                                   * 0.3.9     2025-10-23 [1] Github (fmicompbio/footprintR@b2aef0b)
 generics                                     * 0.1.4     2025-05-09 [1] CRAN (R 4.5.0)
 GenomeInfoDb                                 * 1.44.3    2025-09-21 [1] Bioconductor 3.21 (R 4.5.0)
 GenomeInfoDbData                               1.2.14    2025-04-17 [1] Bioconductor
 GenomicAlignments                              1.44.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 GenomicFeatures                              * 1.60.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 GenomicRanges                                * 1.60.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 ggforce                                        0.5.0     2025-06-18 [1] CRAN (R 4.5.0)
 ggplot2                                      * 4.0.0     2025-09-11 [1] CRAN (R 4.5.0)
 glue                                           1.8.0     2024-09-30 [1] CRAN (R 4.5.0)
 gtable                                         0.3.6     2024-10-25 [1] CRAN (R 4.5.0)
 htmltools                                      0.5.8.1   2024-04-04 [1] CRAN (R 4.5.0)
 htmlwidgets                                    1.6.4     2023-12-06 [1] CRAN (R 4.5.0)
 httr                                           1.4.7     2023-08-15 [1] CRAN (R 4.5.0)
 IRanges                                      * 2.42.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 jsonlite                                       2.0.0     2025-03-27 [1] CRAN (R 4.5.0)
 KEGGREST                                       1.48.1    2025-06-22 [1] Bioconductor 3.21 (R 4.5.0)
 knitr                                          1.50      2025-03-16 [1] CRAN (R 4.5.0)
 labeling                                       0.4.3     2023-08-29 [1] CRAN (R 4.5.0)
 lattice                                        0.22-7    2025-04-02 [1] CRAN (R 4.5.0)
 lifecycle                                      1.0.4     2023-11-07 [1] CRAN (R 4.5.0)
 magrittr                                       2.0.4     2025-09-12 [1] CRAN (R 4.5.0)
 MASS                                           7.3-65    2025-02-28 [2] CRAN (R 4.5.1)
 Matrix                                         1.7-3     2025-03-11 [2] CRAN (R 4.5.1)
 MatrixGenerics                                 1.20.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 matrixStats                                    1.5.0     2025-01-07 [1] CRAN (R 4.5.0)
 memoise                                        2.0.1     2021-11-26 [1] CRAN (R 4.5.0)
 patchwork                                    * 1.3.2     2025-08-25 [1] CRAN (R 4.5.1)
 pillar                                         1.11.1    2025-09-17 [1] CRAN (R 4.5.0)
 pkgconfig                                      2.0.3     2019-09-22 [1] CRAN (R 4.5.0)
 png                                            0.1-8     2022-11-29 [1] CRAN (R 4.5.0)
 polyclip                                       1.10-7    2024-07-23 [1] CRAN (R 4.5.0)
 purrr                                          1.1.0     2025-07-10 [1] CRAN (R 4.5.0)
 R6                                             2.6.1     2025-02-15 [1] CRAN (R 4.5.0)
 RColorBrewer                                   1.1-3     2022-04-03 [1] CRAN (R 4.5.0)
 Rcpp                                           1.1.0     2025-07-02 [1] CRAN (R 4.5.0)
 RCurl                                          1.98-1.17 2025-03-22 [1] CRAN (R 4.5.0)
 restfulr                                       0.0.16    2025-06-27 [1] CRAN (R 4.5.0)
 rjson                                          0.2.23    2024-09-16 [1] CRAN (R 4.5.0)
 rlang                                          1.1.6     2025-04-11 [1] CRAN (R 4.5.0)
 rmarkdown                                      2.30      2025-09-28 [1] CRAN (R 4.5.0)
 Rsamtools                                      2.24.1    2025-09-07 [1] Bioconductor 3.21 (R 4.5.0)
 RSQLite                                        2.4.3     2025-08-20 [1] CRAN (R 4.5.1)
 rtracklayer                                  * 1.68.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 S4Arrays                                       1.8.1     2025-06-01 [1] Bioconductor 3.21 (R 4.5.0)
 S4Vectors                                    * 0.46.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 S7                                             0.2.0     2024-11-07 [1] CRAN (R 4.5.0)
 scales                                         1.4.0     2025-04-24 [1] CRAN (R 4.5.0)
 sessioninfo                                    1.2.3     2025-02-05 [1] CRAN (R 4.5.0)
 SparseArray                                    1.8.1     2025-07-23 [1] Bioconductor 3.21 (R 4.5.0)
 SummarizedExperiment                           1.38.1    2025-04-30 [1] Bioconductor 3.21 (R 4.5.0)
 tibble                                         3.3.0     2025-06-08 [1] CRAN (R 4.5.0)
 tidyr                                          1.3.1     2024-01-24 [1] CRAN (R 4.5.0)
 tidyselect                                     1.2.1     2024-03-11 [1] CRAN (R 4.5.0)
 tweenr                                         2.0.3     2024-02-26 [1] CRAN (R 4.5.0)
 TxDb.Mmusculus.GENCODE.GRCm39.gencodeM34     * 0.1.0     2025-04-17 [1] Bioconductor
 UCSC.utils                                     1.4.0     2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 vctrs                                          0.6.5     2023-12-01 [1] CRAN (R 4.5.0)
 viridisLite                                    0.4.2     2023-05-02 [1] CRAN (R 4.5.0)
 withr                                          3.0.2     2024-10-28 [1] CRAN (R 4.5.0)
 xfun                                           0.53      2025-08-19 [1] CRAN (R 4.5.1)
 XML                                            3.99-0.19 2025-08-22 [1] CRAN (R 4.5.1)
 XVector                                      * 0.48.0    2025-04-15 [1] Bioconductor 3.21 (R 4.5.0)
 yaml                                           2.3.10    2024-07-26 [1] CRAN (R 4.5.0)
 zoo                                            1.8-14    2025-04-10 [1] CRAN (R 4.5.0)

 [1] /tungstenfs/groups/gbioinfo/Appz/R/BioC/R-4.5-release-foss-2024.05_BioC-3.21-release-foss-2024.05
 [2] /tachyon/groups/gbioinfo/Appz/easybuild/software/R/4.5.1-foss-2024.05/lib64/R/library
 * ── Packages attached to the search path.

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</details>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/fmicompbio\.github\.io\/footprintR_book");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./qc-filtering.html" class="pagination-link" aria-label="Quality control and filtering">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Quality control and filtering</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./nucleosomes.html" class="pagination-link" aria-label="Nucleosome analyses">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Nucleosome analyses</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb23" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Visualizing regions with `plotRegion` {#sec-visualize-regions}</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} function in <span class="co">[</span><span class="ot">fmicompbio/footprintR</span><span class="co">]</span>{.githubpkg} provides a flexible interface for plotting data in a given genomic region. </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>As will be illustrated below, both read-level data, summary-level data (aggregating across all reads overlapping a position) and data from BigWig files can visualized. </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>In addition, predefined regions can be highlighted with background color in any track, and gene models or any other genomic region indicators can be added. </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>Generally, <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} builds up a plot as a combination of individual _tracks_, of different _types_ (representing the different data types mentioned above).</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>To add a track to a plot, at least the corresponding <span class="in">`trackType`</span> and <span class="in">`trackData`</span> arguments must be specified. </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>A list of the supported track types is available in the documentation of <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} (see the <span class="in">`tracks`</span> argument).</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>The order and content of the tracks are fully customizable by the user. </span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preparation</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>We start by loading the required packages. </span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>In addition to the software package, we load a <span class="in">`BSgenome`</span> and <span class="in">`TxDb`</span> object providing the mouse genome sequence and transcript annotation.</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-packages</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>BSgenomeName <span class="ot">&lt;-</span> <span class="st">"BSgenome.Mmusculus.GENCODE.GRCm39.gencodeM34"</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>TxDbName <span class="ot">&lt;-</span> <span class="st">"TxDb.Mmusculus.GENCODE.GRCm39.gencodeM34"</span></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(footprintR)</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenomeName, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TxDbName, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Load genome</span></span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>gnm <span class="ot">&lt;-</span> <span class="fu">get</span>(BSgenomeName)</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(gnm) <span class="ot">&lt;-</span> <span class="st">"mm39"</span></span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Load TxDb and create GRangesList with transcript annotations</span></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> <span class="fu">get</span>(TxDbName)</span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a>ebt <span class="ot">&lt;-</span> <span class="fu">exonsBy</span>(txdb, <span class="st">"tx"</span>, <span class="at">use.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-43"><a href="#cb23-43" aria-hidden="true" tabindex="-1"></a><span class="fu">## Read data</span></span>
<span id="cb23-44"><a href="#cb23-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-45"><a href="#cb23-45" aria-hidden="true" tabindex="-1"></a>To exemplify the plotting capabilities of <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}, we use <span class="in">`modBam`</span> files from two wild-type samples, for which 6mA modification calling has been performed. </span>
<span id="cb23-46"><a href="#cb23-46" aria-hidden="true" tabindex="-1"></a>We use the <span class="co">[</span><span class="ot">readModBam</span><span class="co">]</span>{.fn} function to read data from an 800-bp region on chromosome 8 and store it in a <span class="in">`SummarizedExperiment`</span> object. </span>
<span id="cb23-47"><a href="#cb23-47" aria-hidden="true" tabindex="-1"></a>For more information about reading data with <span class="co">[</span><span class="ot">fmicompbio/footprintR</span><span class="co">]</span>{.githubpkg}, see @sec-reading-data. </span>
<span id="cb23-48"><a href="#cb23-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-51"><a href="#cb23-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-52"><a href="#cb23-52" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: read-data</span></span>
<span id="cb23-53"><a href="#cb23-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-54"><a href="#cb23-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data </span></span>
<span id="cb23-55"><a href="#cb23-55" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(</span>
<span id="cb23-56"><a href="#cb23-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">bamfiles =</span> <span class="fu">c</span>(<span class="at">wt1 =</span> <span class="st">"data/mESC_wt_6mA_rep1.bam"</span>,</span>
<span id="cb23-57"><a href="#cb23-57" aria-hidden="true" tabindex="-1"></a>                 <span class="at">wt2 =</span> <span class="st">"data/mESC_wt_6mA_rep2.bam"</span>),</span>
<span id="cb23-58"><a href="#cb23-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb23-59"><a href="#cb23-59" aria-hidden="true" tabindex="-1"></a>    <span class="at">regions =</span> <span class="st">"chr8:39286301-39287100"</span>, </span>
<span id="cb23-60"><a href="#cb23-60" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), </span>
<span id="cb23-61"><a href="#cb23-61" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb23-62"><a href="#cb23-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceReference =</span> gnm</span>
<span id="cb23-63"><a href="#cb23-63" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-64"><a href="#cb23-64" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-65"><a href="#cb23-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-66"><a href="#cb23-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-67"><a href="#cb23-67" aria-hidden="true" tabindex="-1"></a><span class="fu">## Read-level plots</span></span>
<span id="cb23-68"><a href="#cb23-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-69"><a href="#cb23-69" aria-hidden="true" tabindex="-1"></a>Read-level plots can be generated using either the <span class="in">`Heatmap`</span> or the <span class="in">`Lollipop`</span> track type, assuming that there is at least one assay with read-level data present in the provided <span class="in">`SummarizedExperiment`</span> object.</span>
<span id="cb23-70"><a href="#cb23-70" aria-hidden="true" tabindex="-1"></a>The heatmap visualizes each base with a modification call as a filled rectangle, with optional interpolation between observations. </span>
<span id="cb23-71"><a href="#cb23-71" aria-hidden="true" tabindex="-1"></a>The lollipop plot visualizes each base with a modification call as a filled circle. </span>
<span id="cb23-72"><a href="#cb23-72" aria-hidden="true" tabindex="-1"></a>In both cases, the fill color represents the recorded modification probability of the base. </span>
<span id="cb23-73"><a href="#cb23-73" aria-hidden="true" tabindex="-1"></a>For the read-level plot tracks, the <span class="in">`trackData`</span> argument specifies the name of the (read-level) assay that should be displayed.</span>
<span id="cb23-74"><a href="#cb23-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-77"><a href="#cb23-77" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-78"><a href="#cb23-78" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-79"><a href="#cb23-79" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-80"><a href="#cb23-80" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level</span></span>
<span id="cb23-81"><a href="#cb23-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-82"><a href="#cb23-82" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-83"><a href="#cb23-83" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-84"><a href="#cb23-84" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>, </span>
<span id="cb23-85"><a href="#cb23-85" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-86"><a href="#cb23-86" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>),</span>
<span id="cb23-87"><a href="#cb23-87" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>)</span>
<span id="cb23-88"><a href="#cb23-88" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb23-89"><a href="#cb23-89" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-90"><a href="#cb23-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-91"><a href="#cb23-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-92"><a href="#cb23-92" aria-hidden="true" tabindex="-1"></a>In the current plot, reads on both strands are visualized, and can be distinguished based on the position of the modified bases (the As).</span>
<span id="cb23-93"><a href="#cb23-93" aria-hidden="true" tabindex="-1"></a>It is possible to visualize only the reads on one strand, by providing a stranded region to <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}.</span>
<span id="cb23-94"><a href="#cb23-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-97"><a href="#cb23-97" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-98"><a href="#cb23-98" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 6</span></span>
<span id="cb23-99"><a href="#cb23-99" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-100"><a href="#cb23-100" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-stranded</span></span>
<span id="cb23-101"><a href="#cb23-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-102"><a href="#cb23-102" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-103"><a href="#cb23-103" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-104"><a href="#cb23-104" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800:+"</span>, </span>
<span id="cb23-105"><a href="#cb23-105" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-106"><a href="#cb23-106" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>),</span>
<span id="cb23-107"><a href="#cb23-107" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>)</span>
<span id="cb23-108"><a href="#cb23-108" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb23-109"><a href="#cb23-109" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-110"><a href="#cb23-110" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-111"><a href="#cb23-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-112"><a href="#cb23-112" aria-hidden="true" tabindex="-1"></a>Moreover, we can limit the display to only positions where the annotated genome has a given sequence (here, an A).</span>
<span id="cb23-113"><a href="#cb23-113" aria-hidden="true" tabindex="-1"></a>This is often helpful to minimize the impact of base calling errors. </span>
<span id="cb23-114"><a href="#cb23-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-117"><a href="#cb23-117" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-118"><a href="#cb23-118" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 6</span></span>
<span id="cb23-119"><a href="#cb23-119" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-120"><a href="#cb23-120" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-seqcontext</span></span>
<span id="cb23-121"><a href="#cb23-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-122"><a href="#cb23-122" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-123"><a href="#cb23-123" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-124"><a href="#cb23-124" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800:+"</span>, </span>
<span id="cb23-125"><a href="#cb23-125" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-126"><a href="#cb23-126" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>),</span>
<span id="cb23-127"><a href="#cb23-127" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>)</span>
<span id="cb23-128"><a href="#cb23-128" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-129"><a href="#cb23-129" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-130"><a href="#cb23-130" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-131"><a href="#cb23-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-132"><a href="#cb23-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-133"><a href="#cb23-133" aria-hidden="true" tabindex="-1"></a>Each plot function comes with a set of arguments that can be used to control the way that data is displayed in the track (see the documentation of <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} for the full list).</span>
<span id="cb23-134"><a href="#cb23-134" aria-hidden="true" tabindex="-1"></a>For example, we can add interpolation to our heatmap, disable the automatic clustering of the reads, and add custom track and legend titles. </span>
<span id="cb23-135"><a href="#cb23-135" aria-hidden="true" tabindex="-1"></a>For the lollipop plot, we can similarly change the size of the circles and the linewidth of the circle outline.</span>
<span id="cb23-136"><a href="#cb23-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-139"><a href="#cb23-139" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-140"><a href="#cb23-140" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-141"><a href="#cb23-141" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-142"><a href="#cb23-142" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-interpolate</span></span>
<span id="cb23-143"><a href="#cb23-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-144"><a href="#cb23-144" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-145"><a href="#cb23-145" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-146"><a href="#cb23-146" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-147"><a href="#cb23-147" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-148"><a href="#cb23-148" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-149"><a href="#cb23-149" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-150"><a href="#cb23-150" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-151"><a href="#cb23-151" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-152"><a href="#cb23-152" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-153"><a href="#cb23-153" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-154"><a href="#cb23-154" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-155"><a href="#cb23-155" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-156"><a href="#cb23-156" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-157"><a href="#cb23-157" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-158"><a href="#cb23-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-159"><a href="#cb23-159" aria-hidden="true" tabindex="-1"></a>Additional customization options for read-level plots include the possibility to change the color scheme, set the order of the reads, and restricting the set of visualized reads to those that cover a certain fraction of the plot region. </span>
<span id="cb23-160"><a href="#cb23-160" aria-hidden="true" tabindex="-1"></a>Below, we use a 'inferno' color scale for the heatmap, and a gradient from white to blue for the lollipop plot. </span>
<span id="cb23-161"><a href="#cb23-161" aria-hidden="true" tabindex="-1"></a>In addition, we order the reads by the average modification fraction across the region (<span class="in">`orderReads = "regionAvg"`</span>), and exclude any reads that don't cover at least 95% of the plotted region (<span class="in">`minCoveredFraction = 0.95`</span>). </span>
<span id="cb23-162"><a href="#cb23-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-165"><a href="#cb23-165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-166"><a href="#cb23-166" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-167"><a href="#cb23-167" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-168"><a href="#cb23-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-colors</span></span>
<span id="cb23-169"><a href="#cb23-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-170"><a href="#cb23-170" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-171"><a href="#cb23-171" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-172"><a href="#cb23-172" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-173"><a href="#cb23-173" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb23-174"><a href="#cb23-174" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-175"><a href="#cb23-175" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-176"><a href="#cb23-176" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-177"><a href="#cb23-177" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb23-178"><a href="#cb23-178" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-179"><a href="#cb23-179" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-180"><a href="#cb23-180" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb23-181"><a href="#cb23-181" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb23-182"><a href="#cb23-182" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-183"><a href="#cb23-183" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-184"><a href="#cb23-184" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-185"><a href="#cb23-185" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-186"><a href="#cb23-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-187"><a href="#cb23-187" aria-hidden="true" tabindex="-1"></a>In the plot above, the reads were ordered by the average signal across the entire plotted region. </span>
<span id="cb23-188"><a href="#cb23-188" aria-hidden="true" tabindex="-1"></a>It is also possible to order (or cluster) reads based on the signal in a smaller region, by specifying the <span class="in">`orderRegion`</span> argument. </span>
<span id="cb23-189"><a href="#cb23-189" aria-hidden="true" tabindex="-1"></a>As an example, we order the reads by the average signal across only the last 20 nucleotides of the plotted region.</span>
<span id="cb23-190"><a href="#cb23-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-193"><a href="#cb23-193" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-194"><a href="#cb23-194" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-195"><a href="#cb23-195" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-196"><a href="#cb23-196" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-colors-orderregion</span></span>
<span id="cb23-197"><a href="#cb23-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-198"><a href="#cb23-198" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-199"><a href="#cb23-199" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-200"><a href="#cb23-200" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-201"><a href="#cb23-201" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb23-202"><a href="#cb23-202" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-203"><a href="#cb23-203" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-204"><a href="#cb23-204" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb23-205"><a href="#cb23-205" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-206"><a href="#cb23-206" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-207"><a href="#cb23-207" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb23-208"><a href="#cb23-208" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-209"><a href="#cb23-209" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb23-210"><a href="#cb23-210" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-211"><a href="#cb23-211" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-212"><a href="#cb23-212" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb23-213"><a href="#cb23-213" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb23-214"><a href="#cb23-214" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-215"><a href="#cb23-215" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-216"><a href="#cb23-216" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-217"><a href="#cb23-217" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-218"><a href="#cb23-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-219"><a href="#cb23-219" aria-hidden="true" tabindex="-1"></a>For any ordering, we can also reverse the display order of the reads by setting <span class="in">`orderReverse = TRUE`</span>. </span>
<span id="cb23-220"><a href="#cb23-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-223"><a href="#cb23-223" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-224"><a href="#cb23-224" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-225"><a href="#cb23-225" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-226"><a href="#cb23-226" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-colors-orderregion-orderreverse</span></span>
<span id="cb23-227"><a href="#cb23-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-228"><a href="#cb23-228" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-229"><a href="#cb23-229" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-230"><a href="#cb23-230" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-231"><a href="#cb23-231" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb23-232"><a href="#cb23-232" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-233"><a href="#cb23-233" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-234"><a href="#cb23-234" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb23-235"><a href="#cb23-235" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-236"><a href="#cb23-236" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">TRUE</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-237"><a href="#cb23-237" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb23-238"><a href="#cb23-238" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-239"><a href="#cb23-239" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb23-240"><a href="#cb23-240" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-241"><a href="#cb23-241" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">TRUE</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-242"><a href="#cb23-242" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb23-243"><a href="#cb23-243" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb23-244"><a href="#cb23-244" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-245"><a href="#cb23-245" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-246"><a href="#cb23-246" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-247"><a href="#cb23-247" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-248"><a href="#cb23-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-249"><a href="#cb23-249" aria-hidden="true" tabindex="-1"></a>The read-level plots will display all reads overlapping the displayed region. </span>
<span id="cb23-250"><a href="#cb23-250" aria-hidden="true" tabindex="-1"></a>In cases where there are too many reads to display, we recommend subsetting the <span class="in">`SummarizedExperiment`</span> object before calling <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}. </span>
<span id="cb23-251"><a href="#cb23-251" aria-hidden="true" tabindex="-1"></a>This can be done conveniently using <span class="co">[</span><span class="ot">subsetReads</span><span class="co">]</span>{.fn}, which allows you to subset the object keeping either a pre-specified set of reads or a random subset. </span>
<span id="cb23-252"><a href="#cb23-252" aria-hidden="true" tabindex="-1"></a>Here, we illustrate this workflow by subsetting the object to five randomly selected reads per sample before calling <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}.</span>
<span id="cb23-253"><a href="#cb23-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-256"><a href="#cb23-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-257"><a href="#cb23-257" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-258"><a href="#cb23-258" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-259"><a href="#cb23-259" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-colors-orderregion-randomsubset</span></span>
<span id="cb23-260"><a href="#cb23-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-261"><a href="#cb23-261" aria-hidden="true" tabindex="-1"></a><span class="co"># First subset the object to the plotted region and remove reads not </span></span>
<span id="cb23-262"><a href="#cb23-262" aria-hidden="true" tabindex="-1"></a><span class="co"># overlapping that region, to restrict the random sampling to reads </span></span>
<span id="cb23-263"><a href="#cb23-263" aria-hidden="true" tabindex="-1"></a><span class="co"># overlapping the plot region</span></span>
<span id="cb23-264"><a href="#cb23-264" aria-hidden="true" tabindex="-1"></a>sesub <span class="ot">&lt;-</span> <span class="fu">filterPositions</span>(se, <span class="at">regions =</span> <span class="st">"chr8:39286600-39286800"</span>, <span class="at">filters =</span> <span class="st">"regions"</span>)</span>
<span id="cb23-265"><a href="#cb23-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-266"><a href="#cb23-266" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample five random reads for each sample</span></span>
<span id="cb23-267"><a href="#cb23-267" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>L)</span>
<span id="cb23-268"><a href="#cb23-268" aria-hidden="true" tabindex="-1"></a>sesub <span class="ot">&lt;-</span> <span class="fu">subsetReads</span>(sesub, <span class="at">randomSubset =</span> <span class="dv">5</span>)</span>
<span id="cb23-269"><a href="#cb23-269" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-270"><a href="#cb23-270" aria-hidden="true" tabindex="-1"></a>    sesub, </span>
<span id="cb23-271"><a href="#cb23-271" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-272"><a href="#cb23-272" aria-hidden="true" tabindex="-1"></a>    <span class="at">minCoveredFraction =</span> <span class="fl">0.95</span>,</span>
<span id="cb23-273"><a href="#cb23-273" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-274"><a href="#cb23-274" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-275"><a href="#cb23-275" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb23-276"><a href="#cb23-276" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-277"><a href="#cb23-277" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">FALSE</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-278"><a href="#cb23-278" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">fillColors =</span> <span class="st">"-inferno"</span>),</span>
<span id="cb23-279"><a href="#cb23-279" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-280"><a href="#cb23-280" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="st">"regionAvg"</span>, </span>
<span id="cb23-281"><a href="#cb23-281" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderRegion =</span> <span class="fu">as</span>(<span class="st">"chr8:39286780-39286800"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-282"><a href="#cb23-282" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReverse =</span> <span class="cn">FALSE</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-283"><a href="#cb23-283" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>, </span>
<span id="cb23-284"><a href="#cb23-284" aria-hidden="true" tabindex="-1"></a>             <span class="at">fillColors =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"blue"</span>))</span>
<span id="cb23-285"><a href="#cb23-285" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-286"><a href="#cb23-286" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-287"><a href="#cb23-287" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-288"><a href="#cb23-288" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-289"><a href="#cb23-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-290"><a href="#cb23-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-291"><a href="#cb23-291" aria-hidden="true" tabindex="-1"></a>By default, read-level tracks are facetted by sample (the columns of <span class="in">`se`</span>).</span>
<span id="cb23-292"><a href="#cb23-292" aria-hidden="true" tabindex="-1"></a>We can facet by an arbitrary column of <span class="in">`colData(se)`</span> by changing the value of the <span class="in">`facetBy`</span> argument for the track, or set the argument to <span class="in">`NULL`</span> to disable the facetting and show reads from all samples together. </span>
<span id="cb23-293"><a href="#cb23-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-296"><a href="#cb23-296" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-297"><a href="#cb23-297" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-298"><a href="#cb23-298" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-299"><a href="#cb23-299" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-nofacet</span></span>
<span id="cb23-300"><a href="#cb23-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-301"><a href="#cb23-301" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-302"><a href="#cb23-302" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-303"><a href="#cb23-303" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-304"><a href="#cb23-304" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-305"><a href="#cb23-305" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-306"><a href="#cb23-306" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-307"><a href="#cb23-307" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">facetBy =</span> <span class="cn">NULL</span>),</span>
<span id="cb23-308"><a href="#cb23-308" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-309"><a href="#cb23-309" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-310"><a href="#cb23-310" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>,</span>
<span id="cb23-311"><a href="#cb23-311" aria-hidden="true" tabindex="-1"></a>             <span class="at">facetBy =</span> <span class="st">"modbase"</span>)</span>
<span id="cb23-312"><a href="#cb23-312" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-313"><a href="#cb23-313" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-314"><a href="#cb23-314" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-315"><a href="#cb23-315" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-316"><a href="#cb23-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-317"><a href="#cb23-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-318"><a href="#cb23-318" aria-hidden="true" tabindex="-1"></a>By default, the x-axis of the plots represent genomic coordinates. </span>
<span id="cb23-319"><a href="#cb23-319" aria-hidden="true" tabindex="-1"></a>In some cases (especially if the modified bases are unevenly distributed), the plot may be easier to read if only the modified bases are displayed, evenly spaced along the x-axis. </span>
<span id="cb23-320"><a href="#cb23-320" aria-hidden="true" tabindex="-1"></a>This can be achieved by setting <span class="in">`modbaseSpace = TRUE`</span> in <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}.</span>
<span id="cb23-321"><a href="#cb23-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-324"><a href="#cb23-324" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-325"><a href="#cb23-325" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-326"><a href="#cb23-326" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-327"><a href="#cb23-327" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-modbasespace</span></span>
<span id="cb23-328"><a href="#cb23-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-329"><a href="#cb23-329" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-330"><a href="#cb23-330" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-331"><a href="#cb23-331" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-332"><a href="#cb23-332" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-333"><a href="#cb23-333" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-334"><a href="#cb23-334" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">FALSE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-335"><a href="#cb23-335" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-336"><a href="#cb23-336" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-337"><a href="#cb23-337" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-338"><a href="#cb23-338" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-339"><a href="#cb23-339" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-340"><a href="#cb23-340" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb23-341"><a href="#cb23-341" aria-hidden="true" tabindex="-1"></a>    <span class="at">modbaseSpace =</span> <span class="cn">TRUE</span></span>
<span id="cb23-342"><a href="#cb23-342" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-343"><a href="#cb23-343" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-344"><a href="#cb23-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-345"><a href="#cb23-345" aria-hidden="true" tabindex="-1"></a>Another alternative is to anchor the x-axis at a given genomic position, and display the coordinates as distances from this anchor point. </span>
<span id="cb23-346"><a href="#cb23-346" aria-hidden="true" tabindex="-1"></a>This is achieved by setting the <span class="in">`referenceCoordinate`</span> argument to the numeric position that should be used as the anchor.</span>
<span id="cb23-347"><a href="#cb23-347" aria-hidden="true" tabindex="-1"></a>Note how the x-axis title changes to reflect the current coordinate system. </span>
<span id="cb23-348"><a href="#cb23-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-351"><a href="#cb23-351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-352"><a href="#cb23-352" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-353"><a href="#cb23-353" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-354"><a href="#cb23-354" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-referencecoordinate</span></span>
<span id="cb23-355"><a href="#cb23-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-356"><a href="#cb23-356" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-357"><a href="#cb23-357" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-358"><a href="#cb23-358" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-359"><a href="#cb23-359" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-360"><a href="#cb23-360" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-361"><a href="#cb23-361" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-362"><a href="#cb23-362" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-363"><a href="#cb23-363" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-364"><a href="#cb23-364" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-365"><a href="#cb23-365" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-366"><a href="#cb23-366" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-367"><a href="#cb23-367" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb23-368"><a href="#cb23-368" aria-hidden="true" tabindex="-1"></a>    <span class="at">referenceCoordinate =</span> <span class="dv">39286700</span></span>
<span id="cb23-369"><a href="#cb23-369" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-370"><a href="#cb23-370" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-371"><a href="#cb23-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-372"><a href="#cb23-372" aria-hidden="true" tabindex="-1"></a>As <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} returns a <span class="in">`patchwork`</span>/<span class="in">`ggplot2`</span> object, even finer control over the plot components is possible. For example, you may want to change the number of ticks on the x axis, which is controlled by <span class="in">`ggplot2`</span>'s <span class="in">`scale_x_continuous()`</span>. Changing the x axis scale will throw a warning because by adding the <span class="in">`scale_x_continuous()`</span>, you are replacing the existing default x axis scale.</span>
<span id="cb23-373"><a href="#cb23-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-374"><a href="#cb23-374" aria-hidden="true" tabindex="-1"></a>Note that if you are plotting multiple tracks with <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}, you need to add <span class="in">`scale_x_continuous()`</span> using the <span class="in">`*`</span> or <span class="in">`&amp;`</span> operator from the <span class="co">[</span><span class="ot">patchwork</span><span class="co">]</span>{.cranpkg} package, instead of <span class="co">[</span><span class="ot">ggplot2</span><span class="co">]</span>{.cranpkg}'s usual <span class="in">`+`</span> operator. This will make sure that the plot element (here the new x axis scale) is added to all tracks instead of just the last one:</span>
<span id="cb23-375"><a href="#cb23-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-378"><a href="#cb23-378" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-379"><a href="#cb23-379" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-380"><a href="#cb23-380" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-381"><a href="#cb23-381" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-read-level-xbreaks</span></span>
<span id="cb23-382"><a href="#cb23-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-383"><a href="#cb23-383" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-384"><a href="#cb23-384" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-385"><a href="#cb23-385" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="st">"chr8:39286600-39286800"</span>,</span>
<span id="cb23-386"><a href="#cb23-386" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-387"><a href="#cb23-387" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-388"><a href="#cb23-388" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Heatmap"</span>,</span>
<span id="cb23-389"><a href="#cb23-389" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-390"><a href="#cb23-390" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Lollipop"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-391"><a href="#cb23-391" aria-hidden="true" tabindex="-1"></a>             <span class="at">orderReads =</span> <span class="cn">NULL</span>, <span class="at">trackTitle =</span> <span class="st">"Lollipop"</span>,</span>
<span id="cb23-392"><a href="#cb23-392" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-393"><a href="#cb23-393" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-394"><a href="#cb23-394" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span>,</span>
<span id="cb23-395"><a href="#cb23-395" aria-hidden="true" tabindex="-1"></a>    <span class="at">referenceCoordinate =</span> <span class="dv">39286700</span></span>
<span id="cb23-396"><a href="#cb23-396" aria-hidden="true" tabindex="-1"></a>) <span class="sc">*</span> <span class="fu">scale_x_continuous</span>(<span class="at">n.breaks =</span> <span class="dv">3</span>)</span>
<span id="cb23-397"><a href="#cb23-397" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-398"><a href="#cb23-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-399"><a href="#cb23-399" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary-level plots</span></span>
<span id="cb23-400"><a href="#cb23-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-401"><a href="#cb23-401" aria-hidden="true" tabindex="-1"></a>In addition to read-level plots, <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} can also display summary-level data (a single value per genomic position and sample), if they are present in the <span class="in">`SummarizedExperiment`</span> object. </span>
<span id="cb23-402"><a href="#cb23-402" aria-hidden="true" tabindex="-1"></a>These values can be displayed as individual points, and/or with a smooth curve. </span>
<span id="cb23-403"><a href="#cb23-403" aria-hidden="true" tabindex="-1"></a>Summary-level assays can be generated from read-level values using, e.g., the <span class="co">[</span><span class="ot">flattenReadLevelAssay</span><span class="co">]</span>{.fn} function. </span>
<span id="cb23-404"><a href="#cb23-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-407"><a href="#cb23-407" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-408"><a href="#cb23-408" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 7</span></span>
<span id="cb23-409"><a href="#cb23-409" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-410"><a href="#cb23-410" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-summary-level</span></span>
<span id="cb23-411"><a href="#cb23-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-412"><a href="#cb23-412" aria-hidden="true" tabindex="-1"></a><span class="co"># Flatten read-level assay</span></span>
<span id="cb23-413"><a href="#cb23-413" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(</span>
<span id="cb23-414"><a href="#cb23-414" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-415"><a href="#cb23-415" aria-hidden="true" tabindex="-1"></a>    <span class="at">assayName =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-416"><a href="#cb23-416" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="fu">c</span>(<span class="st">"Nmod"</span>, <span class="st">"Nvalid"</span>, <span class="st">"FracMod"</span>)</span>
<span id="cb23-417"><a href="#cb23-417" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-418"><a href="#cb23-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-419"><a href="#cb23-419" aria-hidden="true" tabindex="-1"></a><span class="co"># Add summary-level track</span></span>
<span id="cb23-420"><a href="#cb23-420" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-421"><a href="#cb23-421" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-422"><a href="#cb23-422" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39286300-39287100"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-423"><a href="#cb23-423" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-424"><a href="#cb23-424" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-425"><a href="#cb23-425" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb23-426"><a href="#cb23-426" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-427"><a href="#cb23-427" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Smooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>)</span>
<span id="cb23-428"><a href="#cb23-428" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-429"><a href="#cb23-429" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-430"><a href="#cb23-430" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb23-431"><a href="#cb23-431" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb23-432"><a href="#cb23-432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-433"><a href="#cb23-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-434"><a href="#cb23-434" aria-hidden="true" tabindex="-1"></a>In an analogous way to the read-level plots, we can control the display of summary information using the arguments to <span class="in">`PlotSummaryPointSmooth`</span> (see documentation for <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}). </span>
<span id="cb23-435"><a href="#cb23-435" aria-hidden="true" tabindex="-1"></a>For example, we can change the smoothing function and the degree of smoothing. </span>
<span id="cb23-436"><a href="#cb23-436" aria-hidden="true" tabindex="-1"></a>We can also add the individual points (summary values for individual positions) to the summary plot.</span>
<span id="cb23-437"><a href="#cb23-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-440"><a href="#cb23-440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-441"><a href="#cb23-441" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 7</span></span>
<span id="cb23-442"><a href="#cb23-442" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-443"><a href="#cb23-443" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-summary-level-points</span></span>
<span id="cb23-444"><a href="#cb23-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-445"><a href="#cb23-445" aria-hidden="true" tabindex="-1"></a><span class="co"># Add summary-level plot</span></span>
<span id="cb23-446"><a href="#cb23-446" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-447"><a href="#cb23-447" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-448"><a href="#cb23-448" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39286300-39287100"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-449"><a href="#cb23-449" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-450"><a href="#cb23-450" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-451"><a href="#cb23-451" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb23-452"><a href="#cb23-452" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-453"><a href="#cb23-453" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb23-454"><a href="#cb23-454" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb23-455"><a href="#cb23-455" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb23-456"><a href="#cb23-456" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>))</span>
<span id="cb23-457"><a href="#cb23-457" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-458"><a href="#cb23-458" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-459"><a href="#cb23-459" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb23-460"><a href="#cb23-460" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb23-461"><a href="#cb23-461" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-462"><a href="#cb23-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-463"><a href="#cb23-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-464"><a href="#cb23-464" aria-hidden="true" tabindex="-1"></a><span class="fu">## Highlight regions</span></span>
<span id="cb23-465"><a href="#cb23-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-466"><a href="#cb23-466" aria-hidden="true" tabindex="-1"></a>Sometimes it is helpful to be able to highlight specific genomic regions in the plots. </span>
<span id="cb23-467"><a href="#cb23-467" aria-hidden="true" tabindex="-1"></a>These could be, for example, regions that have been identified as harboring an interesting signal. </span>
<span id="cb23-468"><a href="#cb23-468" aria-hidden="true" tabindex="-1"></a>In <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn}, this can be achieved via the <span class="in">`highlightRegions`</span> argument to the respective tracks. </span>
<span id="cb23-469"><a href="#cb23-469" aria-hidden="true" tabindex="-1"></a>The input provided to this argument should be a <span class="in">`GRanges`</span> object. </span>
<span id="cb23-470"><a href="#cb23-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-473"><a href="#cb23-473" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-474"><a href="#cb23-474" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 7</span></span>
<span id="cb23-475"><a href="#cb23-475" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-476"><a href="#cb23-476" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-highlightregions</span></span>
<span id="cb23-477"><a href="#cb23-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-478"><a href="#cb23-478" aria-hidden="true" tabindex="-1"></a><span class="co"># Define regions to highlight</span></span>
<span id="cb23-479"><a href="#cb23-479" aria-hidden="true" tabindex="-1"></a>reg_to_highlight <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb23-480"><a href="#cb23-480" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="st">"chr8"</span>,</span>
<span id="cb23-481"><a href="#cb23-481" aria-hidden="true" tabindex="-1"></a>    <span class="at">ranges =</span> <span class="fu">IRanges</span>(</span>
<span id="cb23-482"><a href="#cb23-482" aria-hidden="true" tabindex="-1"></a>        <span class="at">start =</span> <span class="fu">c</span>(<span class="dv">39286800</span>, <span class="dv">39286950</span>),</span>
<span id="cb23-483"><a href="#cb23-483" aria-hidden="true" tabindex="-1"></a>        <span class="at">end =</span> <span class="fu">c</span>(<span class="dv">39286900</span>, <span class="dv">39287000</span>)</span>
<span id="cb23-484"><a href="#cb23-484" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb23-485"><a href="#cb23-485" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-486"><a href="#cb23-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-487"><a href="#cb23-487" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-488"><a href="#cb23-488" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-489"><a href="#cb23-489" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39286300-39287100"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-490"><a href="#cb23-490" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-491"><a href="#cb23-491" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-492"><a href="#cb23-492" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb23-493"><a href="#cb23-493" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-494"><a href="#cb23-494" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb23-495"><a href="#cb23-495" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb23-496"><a href="#cb23-496" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb23-497"><a href="#cb23-497" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>),</span>
<span id="cb23-498"><a href="#cb23-498" aria-hidden="true" tabindex="-1"></a>             <span class="at">highlightRegions =</span> reg_to_highlight)</span>
<span id="cb23-499"><a href="#cb23-499" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-500"><a href="#cb23-500" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-501"><a href="#cb23-501" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb23-502"><a href="#cb23-502" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb23-503"><a href="#cb23-503" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-504"><a href="#cb23-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-505"><a href="#cb23-505" aria-hidden="true" tabindex="-1"></a><span class="fu">## Annotation tracks</span></span>
<span id="cb23-506"><a href="#cb23-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-507"><a href="#cb23-507" aria-hidden="true" tabindex="-1"></a>To further relate the observed data to existing annotations, <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} also allows the user to add one or several tracks representing genomic regions (e.g., gene models). </span>
<span id="cb23-508"><a href="#cb23-508" aria-hidden="true" tabindex="-1"></a>Here, we exemplify how to generate such a track to visualize annotated transcripts in the displayed region.</span>
<span id="cb23-509"><a href="#cb23-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-512"><a href="#cb23-512" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-513"><a href="#cb23-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 7</span></span>
<span id="cb23-514"><a href="#cb23-514" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-515"><a href="#cb23-515" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-annotation</span></span>
<span id="cb23-516"><a href="#cb23-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-517"><a href="#cb23-517" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data from a different region than above (as the previous one does not </span></span>
<span id="cb23-518"><a href="#cb23-518" aria-hidden="true" tabindex="-1"></a><span class="co"># overlap annotated genes)</span></span>
<span id="cb23-519"><a href="#cb23-519" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readModBam</span>(</span>
<span id="cb23-520"><a href="#cb23-520" aria-hidden="true" tabindex="-1"></a>    <span class="at">bamfiles =</span> <span class="fu">c</span>(<span class="at">wt1 =</span> <span class="st">"data/mESC_wt_6mA_rep1.bam"</span>,</span>
<span id="cb23-521"><a href="#cb23-521" aria-hidden="true" tabindex="-1"></a>                 <span class="at">wt2 =</span> <span class="st">"data/mESC_wt_6mA_rep2.bam"</span>),</span>
<span id="cb23-522"><a href="#cb23-522" aria-hidden="true" tabindex="-1"></a>    <span class="at">modbase =</span> <span class="st">"a"</span>, </span>
<span id="cb23-523"><a href="#cb23-523" aria-hidden="true" tabindex="-1"></a>    <span class="at">regions =</span> <span class="st">"chr8:39126917-39128220"</span>, </span>
<span id="cb23-524"><a href="#cb23-524" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqinfo =</span> <span class="fu">seqinfo</span>(gnm), </span>
<span id="cb23-525"><a href="#cb23-525" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContextWidth =</span> <span class="dv">1</span>, </span>
<span id="cb23-526"><a href="#cb23-526" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceReference =</span> gnm</span>
<span id="cb23-527"><a href="#cb23-527" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-528"><a href="#cb23-528" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">flattenReadLevelAssay</span>(</span>
<span id="cb23-529"><a href="#cb23-529" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-530"><a href="#cb23-530" aria-hidden="true" tabindex="-1"></a>    <span class="at">assayName =</span> <span class="st">"mod_prob"</span>,</span>
<span id="cb23-531"><a href="#cb23-531" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="fu">c</span>(<span class="st">"Nmod"</span>, <span class="st">"Nvalid"</span>, <span class="st">"FracMod"</span>)</span>
<span id="cb23-532"><a href="#cb23-532" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-533"><a href="#cb23-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-534"><a href="#cb23-534" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-535"><a href="#cb23-535" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-536"><a href="#cb23-536" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-537"><a href="#cb23-537" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-538"><a href="#cb23-538" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-539"><a href="#cb23-539" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb23-540"><a href="#cb23-540" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-541"><a href="#cb23-541" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"GenomicRegion"</span>, </span>
<span id="cb23-542"><a href="#cb23-542" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackData =</span> <span class="fu">subsetByOverlaps</span>(ebt, <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>))),</span>
<span id="cb23-543"><a href="#cb23-543" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb23-544"><a href="#cb23-544" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb23-545"><a href="#cb23-545" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb23-546"><a href="#cb23-546" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>))</span>
<span id="cb23-547"><a href="#cb23-547" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-548"><a href="#cb23-548" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-549"><a href="#cb23-549" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb23-550"><a href="#cb23-550" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb23-551"><a href="#cb23-551" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-552"><a href="#cb23-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-553"><a href="#cb23-553" aria-hidden="true" tabindex="-1"></a><span class="fu">## `BigWig` tracks</span></span>
<span id="cb23-554"><a href="#cb23-554" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-555"><a href="#cb23-555" aria-hidden="true" tabindex="-1"></a>Finally, <span class="co">[</span><span class="ot">plotRegion</span><span class="co">]</span>{.fn} allows the user to include tracks defined by arbitrary BigWig files. </span>
<span id="cb23-556"><a href="#cb23-556" aria-hidden="true" tabindex="-1"></a>Here, we illustrate this by adding a track containing the coverage obtained from a CTCF ChIP-seq experiment. </span>
<span id="cb23-557"><a href="#cb23-557" aria-hidden="true" tabindex="-1"></a>We also modify the colors used to display the summary-level data. </span>
<span id="cb23-558"><a href="#cb23-558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-561"><a href="#cb23-561" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-562"><a href="#cb23-562" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.width: 7</span></span>
<span id="cb23-563"><a href="#cb23-563" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 8</span></span>
<span id="cb23-564"><a href="#cb23-564" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-bigwig</span></span>
<span id="cb23-565"><a href="#cb23-565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-566"><a href="#cb23-566" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRegion</span>(</span>
<span id="cb23-567"><a href="#cb23-567" aria-hidden="true" tabindex="-1"></a>    se, </span>
<span id="cb23-568"><a href="#cb23-568" aria-hidden="true" tabindex="-1"></a>    <span class="at">region =</span> <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>), </span>
<span id="cb23-569"><a href="#cb23-569" aria-hidden="true" tabindex="-1"></a>    <span class="at">tracks =</span> <span class="fu">list</span>(</span>
<span id="cb23-570"><a href="#cb23-570" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"Heatmap"</span>, <span class="at">trackData =</span> <span class="st">"mod_prob"</span>, </span>
<span id="cb23-571"><a href="#cb23-571" aria-hidden="true" tabindex="-1"></a>             <span class="at">interpolate =</span> <span class="cn">TRUE</span>, <span class="at">orderReads =</span> <span class="cn">NULL</span>, </span>
<span id="cb23-572"><a href="#cb23-572" aria-hidden="true" tabindex="-1"></a>             <span class="at">legendTitle =</span> <span class="st">"6mA"</span>),</span>
<span id="cb23-573"><a href="#cb23-573" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"GenomicRegion"</span>, </span>
<span id="cb23-574"><a href="#cb23-574" aria-hidden="true" tabindex="-1"></a>             <span class="at">trackData =</span> <span class="fu">subsetByOverlaps</span>(ebt, <span class="fu">as</span>(<span class="st">"chr8:39126917-39128220"</span>, <span class="st">"GRanges"</span>))),</span>
<span id="cb23-575"><a href="#cb23-575" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"BigWig"</span>, <span class="at">trackData =</span> <span class="fu">c</span>(<span class="at">wt =</span> <span class="st">"data/mESC_wt_CTCF_ChIP_rep1.bw"</span>),</span>
<span id="cb23-576"><a href="#cb23-576" aria-hidden="true" tabindex="-1"></a>             <span class="at">yAxisLabel =</span> <span class="st">"CTCF"</span>, <span class="at">showLegend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb23-577"><a href="#cb23-577" aria-hidden="true" tabindex="-1"></a>             <span class="at">colors =</span> <span class="fu">c</span>(<span class="at">wt =</span> <span class="st">"forestgreen"</span>)),</span>
<span id="cb23-578"><a href="#cb23-578" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">trackType =</span> <span class="st">"PointSmooth"</span>, <span class="at">trackData =</span> <span class="st">"FracMod"</span>, </span>
<span id="cb23-579"><a href="#cb23-579" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistPoint =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>),</span>
<span id="cb23-580"><a href="#cb23-580" aria-hidden="true" tabindex="-1"></a>             <span class="at">smoothMethod =</span> <span class="st">"rollingMean"</span>, <span class="at">windowSize =</span> <span class="dv">15</span>,</span>
<span id="cb23-581"><a href="#cb23-581" aria-hidden="true" tabindex="-1"></a>             <span class="at">arglistSmooth =</span> <span class="fu">list</span>(<span class="at">linewidth =</span> <span class="dv">1</span>),</span>
<span id="cb23-582"><a href="#cb23-582" aria-hidden="true" tabindex="-1"></a>             <span class="at">colors =</span> <span class="fu">c</span>(<span class="at">wt1 =</span> <span class="st">"steelblue"</span>, <span class="at">wt2 =</span> <span class="st">"firebrick2"</span>))</span>
<span id="cb23-583"><a href="#cb23-583" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb23-584"><a href="#cb23-584" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequenceContext =</span> <span class="st">"A"</span></span>
<span id="cb23-585"><a href="#cb23-585" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb23-586"><a href="#cb23-586" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="fl">0.5</span>, <span class="dv">1</span>))</span>
<span id="cb23-587"><a href="#cb23-587" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-588"><a href="#cb23-588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-589"><a href="#cb23-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-590"><a href="#cb23-590" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session info</span></span>
<span id="cb23-591"><a href="#cb23-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-592"><a href="#cb23-592" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb23-593"><a href="#cb23-593" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;&lt;b&gt;</span>
<span id="cb23-594"><a href="#cb23-594" aria-hidden="true" tabindex="-1"></a>Click to view session info</span>
<span id="cb23-595"><a href="#cb23-595" aria-hidden="true" tabindex="-1"></a>&lt;/b&gt;&lt;/summary&gt;</span>
<span id="cb23-598"><a href="#cb23-598" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb23-599"><a href="#cb23-599" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="at">info =</span> <span class="st">"packages"</span>)</span>
<span id="cb23-600"><a href="#cb23-600" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb23-601"><a href="#cb23-601" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb23-602"><a href="#cb23-602" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/fmicompbio/footprintR_book/edit/main/visualize-regions.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/fmicompbio/footprintR_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>